---
title: "preliminary-results"
output: pdf_document
---


```{r}
psi_input_dir <- "suppa_analysis/events"
transcript_type <- 
                    'TSL'
                    # 'coding'
                    # 'coding_TSL'
header <- readLines(paste0('suppa_analysis/tpm_expressions_', transcript_type, '.tsv'), n=1)
sample_cols <- strsplit(header, '\t', TRUE)[[1]]
# all_event_files <- list.files(psi_input_dir, pattern = paste0('[^', paste(as_events, collapse = '|'), ']\\.psi$'), full.names = TRUE)
# all_event_dt <- rbindlist(sapply(all_event_files, fread, simplify=FALSE))
# setnames(other_event_dt, 'V1', 'event')
event_res <- sapply(as_events, function(event_type) {
	# load psi values
  event_dt <- fread(paste0(psi_input_dir, '/event_', transcript_type, '_', event_type, '.psi'))
  setnames(event_dt, 'V1', 'event_id')
  # # get all the samples that are also in the metadata and filter
  # event_dt <- event_dt[, c('event', ..samples_with_epi)]
  # # remove all the events where there is only one valid value. They would be discarded later anyway
  # event_dt <- event_dt[apply(event_dt[, ..sample_cols], 1, function(r) sum(is.na(r)) < (length(r) - 1)), ]
  
  # prepare the alternative/upstream/downstream region
  if (event_type == 'SE') {
    event_dt[, c('Event Type', 'seqnames', 'e1s2', 'e2s3', 'strand') := tstrsplit(event_id, ':', fixed=TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, c('e1', 's2', 'e2', 's3') := c(tstrsplit(e1s2, '-', fixed = TRUE, type.convert = TRUE), tstrsplit(e2s3, '-', fixed = TRUE, type.convert = TRUE))]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = e1, end = s3), strand = strand)]
    upstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', e1, e2) + 1, end = ifelse(strand == '+', s2, s3) - 1), strand = strand)]
    downstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', e2, e1) + 1, end = ifelse(strand == '+', s3, s2) - 1), strand = strand)]
    event_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = s2, end = e2), strand = strand)]
  } else if (event_type == 'RI') {
    event_dt[, c('Event Type', 'seqnames', 's1', 'e1s2', 'e2', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, c('e1', 's2') := tstrsplit(e1s2, '-', fixed = TRUE, type.convert = TRUE)]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = s1, end = e2), strand = strand)]
    upstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', s1, s2), end = ifelse(strand == '+', e2, e2)), strand = strand)]
    downstream_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = ifelse(strand == '+', s2, s1), end = ifelse(strand == '+', e2, e1)), strand = strand)]
    event_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = e1 + 1, end = s2 - 1), strand = strand)]
  } else if (event_type %in% c('AL', 'AF')){
    event_dt[, c('Event Type', 'seqnames', 's1/e1s2', 'e1s3/e2', 's2/e1s3', 'e2s3/e3', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[ifelse(event_type=='AF', strand == '+', strand == '-'), start_all:=as.integer(`s1/e1s2`)]
    event_dt[ifelse(event_type=='AF', strand == '-', strand == '+'), start_all:=tstrsplit(`s1/e1s2`, '-', fixed=TRUE, type.convert = TRUE)[[1]] + 1]
    event_dt[ifelse(event_type=='AF', strand == '-', strand == '+'), end_all:=as.integer(`e2s3/e3`)]
    event_dt[ifelse(event_type=='AF', strand == '+', strand == '-'), end_all:=tstrsplit(`e2s3/e3`, '-', fixed=TRUE, type.convert = TRUE)[[2]] - 1]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = start_all, end = end_all), strand = strand)]
    upstream_gr <- NULL
    downstream_gr <- NULL
    event_gr <- NULL
  } else if (event_type %in% c('A3', 'A5')){
    event_dt[, c('Event Type', 'seqnames', 'e1s2/e2s3', 'e1s3', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, c('e1/e2', 's2/s3') := tstrsplit(`e1s2/e2s3`, '-', fixed = TRUE, type.convert = TRUE)]
    event_dt[, c('e1', 's3') := tstrsplit(e1s3, '-', fixed = TRUE, type.convert = TRUE)]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = `e1/e2` + 1, end = s3 - 1), strand = strand)]
    upstream_gr <- NULL
    downstream_gr <- NULL
    event_gr <- NULL
  } else if (event_type == 'MX'){
    event_dt[, c('Event Type', 'seqnames', 'e1s2', 'e2s4', 'e1s3', 'e3s4', 'strand') := tstrsplit(event_id, ':', fixed=TRUE, type.convert = TRUE)]
    event_dt[, c('gene_id', 'event_type') := tstrsplit(`Event Type`, ';', fixed=T)]
    event_dt[, start_all:=tstrsplit(e1s2, '-', fixed=TRUE, type.convert = TRUE)[[1]] + 1]
    event_dt[, end_all:=tstrsplit(e3s4, '-', fixed=TRUE, type.convert = TRUE)[[2]] - 1]
    all_gr <- event_dt[, GRanges(seqnames = seqnames, ranges = IRanges(start = start_all, end = end_all), strand = strand)]
    upstream_gr <- NULL
    downstream_gr <- NULL
    event_gr <- NULL
  } else stop('unknown event')
  
  list(event_dt = event_dt[, c('event_id', 'gene_id', 'seqnames', 'strand', ..sample_cols)], event_gr = event_gr, upstream_gr = upstream_gr, downstream_gr = downstream_gr, all_gr=all_gr)
}, simplify = FALSE)
```

```{r, fig.width=7, fig.height=4}
library(UpSetR)
library(grid)

all_gr <- Reduce(c, lapply(event_res, function(event){event$all_gr}))

# merge event_dts and gr objects. adjust IDs
# max_ID <- 0L
event_dt <- data.table()
event_gr <- GRanges()
upstream_gr <- GRanges()
downstream_gr <- GRanges()
for (this_event in to_analyze) {
  this_dt <- copy(event_res[[this_event]]$event_dt)
  this_event_gr <- event_res[[this_event]]$event_gr
  single_genes_only <- this_dt[!grepl('_', gene_id, fixed = TRUE), which = TRUE]
  not_gonosomes <- this_dt[!seqnames %in% c('chrY', 'chrX'), which = TRUE]
  # var_events[[this_event]] <- sapply(var_events[[this_event]], function(ids) ids + max_ID, simplify = FALSE)
  
  other_event_overlaps <- findOverlaps(this_event_gr, all_gr)
  non_overlapping_other <- other_event_overlaps@from[which(!(duplicated(other_event_overlaps@from) | duplicated(other_event_overlaps@from, fromLast = TRUE)))]
  self_event_overlaps <- findOverlaps(this_event_gr)
  non_overlapping_self <- self_event_overlaps@from[which(!(duplicated(self_event_overlaps@from) | duplicated(self_event_overlaps@from, fromLast = TRUE)))]  
  pdf(file=file.path(plot_dir, paste0(this_event, transcript_type, '_event_intersections.pdf')), width = 7, height = 5, onefile = FALSE) # or other device
  upset <- upset(fromList(list('all'=this_dt[, seq.int(.N)],'no overlap other'=non_overlapping_other, 'no overlap self'=non_overlapping_self, 'single gene'=single_genes_only, 'not gonosomes' = not_gonosomes)), sets.bar.color = "#56B4E9", order.by = "freq")
  print(upset)
  grid.text(paste(this_event, transcript_type), x = 0.65, y=0.95, gp=gpar(fontsize=20)) 
  dev.off()
  print(upset)
  grid.text(paste(this_event, transcript_type), x = 0.65, y=0.95, gp=gpar(fontsize=20)) 
  # keep_rows <- intersect(single_genes_only, non_overlapping_other)
  # add ID
  # this_dt <- this_dt[keep_rows]
  # this_dt[, ID:=seq.int(.N)]
  
  # this_max_ID <- this_dt[, max(ID)]
  
  event_gr <- c(event_gr, this_event_gr)#[keep_rows])
  upstream_gr <- c(upstream_gr, event_res[[this_event]]$upstream_gr)#[keep_rows])
  downstream_gr <- c(downstream_gr, event_res[[this_event]]$downstream_gr)#[keep_rows])
  # this_dt[, ID := ID + max_ID]
  this_dt[, `Event Type` := this_event]
  event_dt <- rbind(event_dt, this_dt)
  # max_ID <- max_ID + this_max_ID
}
event_dt[, ID:=seq.int(.N)]
event_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]

other_event_overlaps <- findOverlaps(event_gr, all_gr)
non_overlapping_other <- other_event_overlaps@from[which(!(duplicated(other_event_overlaps@from) | duplicated(other_event_overlaps@from, fromLast = TRUE)))]

single_genes_only <- event_dt[!grepl('_', gene_id, fixed = TRUE), which = TRUE]
not_gonosomes <- event_dt[!seqnames %in% c('chrY', 'chrX'), which = TRUE]

keep_rows <- intersect(intersect(single_genes_only, non_overlapping_other), not_gonosomes)

# include experimental events manually, to make sure models are built for this:
manual_ids <- c(4019, 4020, 460:463, 310:314, 8201:8202, 1658, 1048:1049, 13896:13897)
# make sure keep_rows and manual_ids are disjoint
stopifnot(length(intersect(keep_rows, manual_ids)) == 0)
keep_rows_manual <- unique(sort(c(keep_rows, manual_ids)))

stopifnot(identical(event_dt[, ID], seq(nrow(event_dt))))# && max_ID == event_dt[, max(ID)])
cols_to_factor <- c('event_id', 'gene_id', 'seqnames', 'strand', 'Event Type')
event_dt[,(cols_to_factor) := lapply(.SD, as.factor), .SDcols = cols_to_factor]
```


```{r, fig.width=7, fig.height=7}
saveRDS(event_res, 'event_res.rds')
rm(event_res)
```

```{r, eval=FALSE}
# plot ecdf
plot_ecdf <- function(mc_dt, interactive = TRUE, pal = RColorBrewer::brewer.pal(12, "Set3")) {
  require(ggplot2)
  stopifnot(!is.null(mc_dt))
  nk <- length(levels(mc_dt$k))
  if(nk > length(pal)){
    pal <- grDevices::colorRampPalette(colors = pal)(nk)
  }
  ggp <-
    ggplot(mc_dt, aes(x = ConsensusIndex, y = CDF, color = k)) + geom_step() + theme_bw()+
    scale_color_manual(values = pal)
  if (interactive)
    ggp <- plotly::ggplotly(ggp)
  return(ggp)
}

add_max_curvature <- function(dt, column, h = 1) {
  # Ensure dt is a data.table
  if (!inherits(dt, "data.table")) {
    stop("dt must be a data.table")
  }
  # Calculate lagged and lead values of the target column
  dt[, paste0(column, "_lagged") := data.table::shift(get(column), type = "lag")]
  dt[, paste0(column, "_lead") := data.table::shift(get(column), type = "lead")]
  
  # Calculate the second derivative
  dt[, second_derivative := (get(paste0(column, "_lead")) - 2 * get(column) + get(paste0(column, "_lagged"))) / (h ^
                                                                                                                   2)]
  
  # Identify the index of the maximum curvature point
  max_curv_index <- which.max(dt$second_derivative)
  
  # Add a column to indicate the maximum curvature point
  dt[, `maximum curvature` := 'no']
  dt[max_curv_index, `maximum curvature` := 'maximum curvature']
  
  # Optionally, clean up by removing the intermediate columns
  dt[, c(paste0(column, "_lagged"),
         paste0(column, "_lead"),
         "second_derivative") := NULL]
  
  return(invisible(dt))
}

# plot delta area
plot_delta_area <- function(mc_dt, interactive = TRUE) {
  stopifnot(!is.null(mc_dt))
  
  mc_dt <- mc_dt[order(k, ConsensusIndex)]
  
  area_under_curve <- mc_dt[, .(AUC = {
    # Calculate the widths (differences in ConsensusIndex) and heights (CDF values)
    widths <- diff(.SD$ConsensusIndex)
    heights <- head(.SD$CDF,-1)
    # Calculate the area of each rectangle and sum them up
    sum(widths * heights)
  }), by = k]
  
  # Compute delta Area
  area_under_curve[, DeltaAUC := c(NA, diff(AUC))]
  area_under_curve[k == '2', DeltaAUC := AUC]
  
  add_max_curvature(area_under_curve, "DeltaAUC")
  
  # View the results
  ggp <-
    ggplot(area_under_curve, aes(x = k, y = DeltaAUC, group = 1)) + geom_line() + geom_point(aes(color =
                                                                                                   `maximum curvature`)) +
    scale_color_manual(values = c('maximum curvature' = 'red')) + theme_bw()
  if (interactive) {
    ggp <- plotly::ggplotly(ggp)
    for (i in seq_along(ggp$x$data))
      if (length(ggp$x$data[[i]]$legendgroup) > 0)
        if (ggp$x$data[[i]]$legendgroup == "no")
          ggp$x$data[[i]]$showlegend <- FALSE
  }
  
  return(ggp)
}

# plot PAC
plot_pac <- function(mc_dt,
                     interactive = TRUE,
                     x1 = .05,
                     x2 = 1 - x1) {
  stopifnot(!is.null(mc_dt))
  
  pac_dt <-
    merge(mc_dt[ConsensusIndex <= x1, .(y1 = max(CDF)), by = k], mc_dt[ConsensusIndex <= x2, .(y2 =
                                                                                                 max(CDF)), by = k])
  pac_dt[, PAC := y2 - y1]
  pac_dt[, `minimum PAC` := 'no']
  pac_dt[PAC == min(PAC), `minimum PAC` := 'minimum PAC']
  
  add_max_curvature(pac_dt, "PAC")
  
  # View the results
  ggp <-
    ggplot(pac_dt, aes(x = k,
                       y = PAC,
                       group = 1)) + geom_line() + geom_point(aes(color = `maximum curvature`, shape = `minimum PAC`)) +
    scale_color_manual(values = c('maximum curvature' = 'red')) + scale_shape_manual(values = c('minimum PAC' = 17, 'no' = 16), breaks= 'minimum PAC') + theme_bw()
  if (interactive) {
    ggp <- plotly::ggplotly(ggp)
    for (i in seq_along(ggp$x$data))
      if (length(ggp$x$data[[i]]$legendgroup) > 0)
        if (ggp$x$data[[i]]$legendgroup == "no")
          ggp$x$data[[i]]$showlegend <- FALSE
  }
  return(ggp)
}
```



```{r}
psi_melt <- melt(event_dt, measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'PSI')
psi_stats <- psi_melt[, .(`NA Percentage` = sum(is.na(PSI))/length(sample_cols),
                                                                                           `Mean PSI` = mean(PSI, na.rm = TRUE),
                                                                                           `PSI SD` = sd(PSI, na.rm = TRUE),
                                                                                           `Adjusted PSI SD` = sqrt(sum((PSI - mean(PSI, na.rm = TRUE))^2, 
                                                                                                                        na.rm = TRUE)/(length(sample_cols) - 1)),
                                                                                           MAD = mad(PSI, na.rm = TRUE),
                                                                                           Entropy = entropy::entropy(entropy::discretize(PSI, 20, range(c(0,1))), 
                                                                                             method = 'ML', unit = 'log2')
                                                                                           ), by=ID]
# merge new columns onto event_dtdistance
event_dt[psi_stats, on='ID', let(`NA Percentage` = `NA Percentage`, `Mean PSI` = `Mean PSI`, `PSI SD` = `PSI SD`, `Adjusted PSI SD` = `Adjusted PSI SD`, MAD = MAD, Entropy = Entropy)]
# compute median Adj SD per event type on IDs in keep_rows
median_adj_sd <- event_dt[ID %in% keep_rows, .(med=median(`Adjusted PSI SD`, na.rm = TRUE), tmp_event_type=`Event Type`), by=`Event Type`][, `Event Type`:=NULL][]

event_dt[ID %in% keep_rows, Variability := ifelse(`Adjusted PSI SD`>=median_adj_sd[`Event Type` == tmp_event_type, med], 'High Variability', 'Low Variability'), by=`Event Type`]
event_dt[ID %in% manual_ids, Variability := ifelse(`Adjusted PSI SD`>=median_adj_sd[`Event Type` == tmp_event_type, med], 'High Variability', 'Low Variability'), by=`Event Type`]

event_dt[, Variability := factor(Variability, levels=c('Low Variability', 'High Variability'))]
```


```{r, fig.width = 12, fig.height = 8}
ggplot(event_dt[ID %in% keep_rows], aes(x = `NA Percentage`, y = `PSI SD`)) + geom_point() + theme_bw() + facet_wrap(~`Event Type`)
ggplot(event_dt[ID %in% keep_rows], aes(x = `NA Percentage`, y = `MAD`)) + geom_point() + theme_bw() + facet_wrap(~`Event Type`)
ggplot(event_dt[ID %in% keep_rows], aes(x = `NA Percentage`, y = `Adjusted PSI SD`)) + geom_point() + theme_bw() + facet_wrap(~`Event Type`)
ggplot(event_dt[ID %in% keep_rows], aes(x = `Adjusted PSI SD`, y = MAD, color = `NA Percentage`)) + geom_point() + theme_bw() + facet_wrap(~`Event Type`) + theme(strip.background = element_rect(fill = 'white')) + scale_color_gradient(low = 'darkblue', high = 'lightgrey')
ggplot(event_dt[ID %in% keep_rows], aes(x = `Adjusted PSI SD`, y = Entropy, color = `NA Percentage`)) + geom_point() + theme_bw() + facet_wrap(~`Event Type`) + theme(strip.background = element_rect(fill = 'white')) + scale_color_gradient(low = 'darkblue', high = 'lightgrey')
ggplot(event_dt[ID %in% keep_rows], aes(y = `Adjusted PSI SD`, x = `Mean PSI`, color = `NA Percentage`)) + geom_point() + theme_bw() + facet_wrap(~`Event Type`) + theme(strip.background = element_rect(fill = 'white')) + scale_color_gradient(low = 'darkblue', high = 'lightgrey') + geom_hline(data=event_dt[ID %in% keep_rows, .(median_sd=median(`Adjusted PSI SD`, na.rm = TRUE)), by = `Event Type`], aes(yintercept = median_sd)) 
```

```{r, eval=FALSE}
require(seriation)
lapply(to_analyze, function(this_event){
  no_na_psi_dt <- na.omit(psi_melt[ID %in% keep_rows & `Event Type` == this_event], cols = 'PSI')
  event_ids <- no_na_psi_dt[, unique(ID)]
  distance_combs <- as.data.table(t(combn(event_ids, 2)))
  distance_combs$distance <- pbmcmapply(distance_combs$V1, distance_combs$V2, FUN = function(x, y) transport::wasserstein1d(no_na_psi_dt[ID == x, PSI], no_na_psi_dt[ID == y, PSI]))
  dist_obj <- as.dist(as.matrix(dcast(rbind(distance_combs, setnames(copy(distance_combs), old = c("V1", "V2"), new=c("V2", "V1"))),
                                      V1 ~ V2, value.var = 'distance'), rownames = 'V1'))
  psi_mat <- as.matrix(dcast(no_na_psi_dt, IHEC ~ ID, value.var = 'PSI'), rownames = 'IHEC')
  # check if there are any missing values of inf in the distance matrix
  if (any(is.na(dist_obj)) | any(is.infinite(dist_obj))) {
    warning("There are missing or infinite values in the distance matrix. Please check the data.")
  }
  # maxK <- 30
  # ccp <- ConsensusClusterPlus::ConsensusClusterPlus(dist_obj, maxK = maxK, reps=10, innerLinkage = 'ward.D2', finalLinkage = 'ward.D2', seed = 1, verbose = TRUE)
  # Kvec <- 2:maxK
  # names(Kvec) <- Kvec
  # mc_dt <- rbindlist(lapply(Kvec, function(i){
  #   consensus_matrix <- ccp[[i]]$consensusMatrix
  #   consensus_values <- consensus_matrix[lower.tri(consensus_matrix)]
  #   ecdf_data <- ecdf(consensus_values)
  #   list(ConsensusIndex = consensus_values, CDF = ecdf_data(consensus_values))
  # }), idcol = 'k')
  # mc_dt[, k:=factor(k, levels=Kvec)]
  # heatmap(psi_mat, Colv = as.dendrogram(reorder(hclust(dist_obj, method = 'ward.D2'), dist=dist_obj, method="OLO")), na.rm=FALSE, scale= 'none')
  pheatmap(psi_mat, cluster_cols = reorder(hclust(dist_obj, method = 'ward.D2'), dist=dist_obj, method="OLO"), cutree_cols = 3, show_colnames = FALSE, show_rownames = FALSE)
})
```

```{r, fig.width = 12, fig.height = 12, eval=FALSE}
var_events <- sapply(to_analyze, function(event) {
  this_event_dt <- event_dt[`Event Type` == event]
  event_mat <- as.matrix(this_event_dt[ID %in% keep_rows, ..sample_cols])
  rownames(event_mat) <- this_event_dt[ID %in% keep_rows, ID]
  event_mat[is.nan(event_mat)] <- NA
  na_count <- apply(event_mat, 1, function(r) sum(is.na(r))) # < (length(r) - 1)))
  na_percentage <- na_count / length(sample_cols)
  
  # check event psi variation
  event_psi_sd <- apply(event_mat, 1, sd, na.rm = TRUE)
  event_psi_mean <- apply(event_mat, 1, mean, na.rm = TRUE)
  event_psi_impute_sd <- apply(t(glmnet::na.replace(t(event_mat), m = event_psi_mean)), 1, sd, na.rm = TRUE)
  event_psi_adj_sd <- sqrt(rowSums((event_mat - rowMeans(event_mat, na.rm = TRUE))^2, na.rm = TRUE)/(length(sample_cols) - 1))
  event_psi_entropy <- apply(event_mat, 1, function(PSI) as.numeric(entropy::entropy(entropy::discretize(PSI, 20, range(c(0,1))), method = 'ML', unit = 'log2')))
  
  stopifnot(all(abs(event_psi_impute_sd - event_psi_adj_sd) < 1e-10, na.rm = TRUE))
  
  # hist(event_psi_sd, 30, main = event)
  sd_mean_dt <- data.table(sd = event_psi_sd, event_psi_mean=event_psi_mean, entropy=event_psi_entropy, na_percentage=na_percentage, sd_impute=event_psi_adj_sd)
  print(ggplot(sd_mean_dt, aes(x = na_percentage, y = sd)) + geom_point() + labs(title=event, x=paste0('na percentage (overall ', length(sample_cols), ')')) + theme_bw())
  print(ggplot(sd_mean_dt, aes(x = na_percentage, y = sd_impute)) + geom_point() + labs(title=event, x=paste0('na percentage (overall ', length(sample_cols), ')')) + theme_bw())
  print(ggplot(sd_mean_dt, aes(x = sd_impute, y = entropy)) + geom_point() + labs(title=event) + theme_bw())
  
  quants <- seq(0, .9, .1)
  var_events <- sapply(quants, function(quant) {
    threshold <- quantile(event_psi_adj_sd, quant, na.rm=TRUE)
    idx <- which(event_psi_adj_sd >= threshold)
    stopifnot(identical(names(idx), rownames(event_mat)[idx]))
    sd_mean_dt[idx, percentage:=as.character(quant)]
    names(idx)
  })
  names(var_events) <- quants
  print(ggplot(sd_mean_dt, aes(x = event_psi_mean, y = sd_impute, color = percentage)) + geom_hline(yintercept = quantile(event_psi_adj_sd, quants, na.rm=TRUE, names=FALSE)) + geom_point() + labs(title=event, x='mean psi') + theme_bw())
  
  
  print(pheatmap(event_mat, show_colnames = FALSE, show_rownames = FALSE, main = event, cluster_rows = FALSE, cluster_cols = FALSE))
  rmv <- apply(event_mat, 1, function(r) sum(is.na(r)) > length(r)/2)
  print(pheatmap(event_mat[intersect(var_events[['0.5']], names(rmv)[!rmv]), ], show_colnames = FALSE, main = paste(event, '0.5 var threshold')))
  var_events
}, simplify = FALSE)
# add Variability column to event_dt
event_dt[, Variability := "Low Variability"]
event_dt[ID %in% c(var_events$SE$`0.5`, var_events$RI$`0.5`), Variability := "High Variability"]
```

```{r}
library(ggplot2)
ggplot(melt(event_dt[ID %in% keep_rows, sapply(sample_cols, function(x) sum(is.nan(get(x)))/.N, simplify = FALSE), by = `Event Type`], measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'na_proportion'), aes(x = `Event Type`, y = na_proportion)) + geom_violin() + labs(y = 'Proportion of NAs in PSI Values per Epigenome', x ='Event Name') + geom_dotplot(binaxis='y', stackdir='center', dotsize=.5) + geom_boxplot(width=0.1) + stat_summary(fun=mean, geom="point", shape=23, size=2) + scale_y_continuous(breaks = seq(0, 1, .1))+ theme_bw()
ggsave(filename = file.path(plot_dir, 'sample_nas.pdf'))
# melt(event_dt, id.vars = c('ID', 'gene_id', 'seqnames', 'strand', 'Event Type'), variable.name = 'IHEC', value.name = 'PSI')
ggplot(melt(event_dt[ID %in% keep_rows], id.vars = c('ID', 'gene_id', 'seqnames', 'strand', 'Event Type'), measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'PSI')[, .(na_proportion=sum(is.nan(PSI))/.N), by = .(ID, `Event Type`)], aes(x = `Event Type`, y = na_proportion)) + geom_violin() + labs(y = 'Proportion of NAs in PSI Values per Event', x ='Event Type')  + geom_dotplot(binaxis='y', stackdir='center', dotsize=.1, binwidth = .01) + geom_boxplot(width=0.1) + stat_summary(fun=mean, geom="point", shape=23, size=2) + scale_y_continuous(breaks = seq(0, 1, .1))+ theme_bw()
ggsave(filename = file.path(plot_dir, 'event_nas.pdf'))
ggplot(melt(event_dt[ID %in% keep_rows], id.vars = c('ID', 'gene_id', 'seqnames', 'strand', 'Event Type'), measure.vars = sample_cols, variable.name = 'IHEC', value.name = 'PSI')[, .(na_proportion=sum(is.nan(PSI))/(.N)), by = .(ID, `Event Type`)], aes(x = na_proportion)) + facet_wrap(. ~ `Event Type`, scales = 'free') + stat_ecdf() + labs(title = 'na proportion by events')
```

```{r, eval=FALSE}
library(Biostrings)
genome_fasta <- 'GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz'
genome <- readDNAStringSet(genome_fasta)
chr_names <- tstrsplit(names(genome), "[ \t]+", keep = 1)[[1]] # split names by tab/space
names(genome) <- chr_names

# make new genomic ranges object from event_dt: create 5' 9mer 9 bases long. [3 bases in exon][6 bases in intron] for SE, we need to take the end here
down5ss <- promoters(downstream_gr, upstream = 3, downstream = 6)
event5ss <- promoters(event_gr, upstream = 3, downstream = 6)
is_se <- event_dt[, `Event Type` == 'SE']
se5ss_ranges <- down5ss[is_se]
ri5ss_ranges <- event5ss[!is_se]
ss5_indices <- order(c(which(is_se), which(!is_se)))
ss5_gr <- c(se5ss_ranges, ri5ss_ranges)[ss5_indices]

ss5 <- genome[ss5_gr]
ss5[event_dt[, strand == '-']] <- reverseComplement(ss5[event_dt[, strand == '-']])
names(ss5) <- event_dt[, ID]
ss5 <- ss5[sapply(vmatchPattern("N", ss5), function(x) length(x) == 0)]
writeXStringSet(ss5, filepath = '5ss.fasta')

# make new genomic ranges object from event_dt: create 3' 23mer 23 bases long. [20 bases in the intron][3 base in the exon] for SE, we need to take the start here
event3ss <- promoters(event_gr, upstream = 20, downstream = 3)
down3ss <- promoters(downstream_gr, upstream = 20, downstream = 3)
se3ss_ranges <- event3ss[is_se]
ri3ss_ranges <- down3ss[!is_se]
ss3_indices <- order(c(which(is_se), which(!is_se)))
ss3_gr <- c(se3ss_ranges, ri3ss_ranges)[ss3_indices]

ss3 <- genome[ss3_gr]
ss3[event_dt[, strand == '-']] <- reverseComplement(ss3[event_dt[, strand == '-']])
names(ss3) <- event_dt[, ID]
ss3 <- ss3[sapply(vmatchPattern("N", ss3), function(x) length(x) == 0)]
writeXStringSet(ss3, filepath = '3ss.fasta')
```

```{bash, eval=FALSE, engine.opts="-l"}
conda activate ihec-as
conda install -y -c conda-forge suppa==5.32
```

```{bash, eval=FALSE, engine.opts="-l"}
conda activate ihec-as
wget http://hollywood.mit.edu/burgelab/maxent/download/fordownload.tar.gz
gunzip fordownload.tar.gz
tar -xf fordownload.tar
```


```{bash, eval=FALSE, engine.opts="-l"}
conda activate ihec-as
cd fordownload/
perl score3.pl ../3ss.fasta > ../3scores.txt
perl score5.pl ../5ss.fasta > ../5scores.txt
```

```{r}
scores3 <- fread('3scores.txt')
names(scores3) <- c('sequence', '3ss')
header_3 <- grep("^>", readLines('3ss.fasta'), value = TRUE)
scores3[, ID:=as.integer(substr(header_3, 2, nchar(header_3)))]
scores3[, sequence:=NULL]

scores5 <- fread('5scores.txt')
names(scores5) <- c('sequence', '5ss')
header_5 <- grep("^>", readLines('5ss.fasta'), value = TRUE)
scores5[, ID:=as.integer(substr(header_5, 2, nchar(header_5)))]
scores5[, sequence:=NULL]

# rm(ss3, ss5, se5ss_ranges, ri5ss_ranges, se3ss_ranges, ri3ss_ranges, is_se, ss5_gr, ss3_gr, ss5_indices, ss3_indices, chr_names, down5ss, down3ss, event5ss, event3ss, genome)
```

```{r}
metadata <- data.table::fread(sample_metadata_file)
metadata <- metadata[epirr_id_without_version %in% sample_cols]
ihec_ia_colors <- unlist(jsonlite::read_json('IHEC_EpiATLAS_IA_colors_Apr_2024_Final.json'), recursive = FALSE)
sample_hex_colors <- sapply(unlist(ihec_ia_colors$fig1_ontology_intermediate_merged, recursive = FALSE), function(x) {
    cols <- as.numeric(strsplit(x, ",")[[1]])
    rgb(cols[1], cols[2], cols[3], maxColorValue = 255)
})
all_files <- fread(file.path(data_dir, 'epiatlas_metadata.csv'))
metadata[all_files[epirr_id_without_version %in% sample_cols & assay_type == "RNA-Seq", .(`mRNA-Seq`='mRNA-Seq' %in% experiment_type), by=epirr_id_without_version], on='epirr_id_without_version', `mRNA-Seq`:=`mRNA-Seq`]
# metadata[, epirr_id_without_version := data.table::tstrsplit(EpiRR, '.', fixed = TRUE)[1]]
# metadata[, annotation := harmonized_sample_ontology_intermediate]
# metadata[biomaterial_type == 'primary tissue', annotation := tissue_type]
# metadata[biomaterial_type == 'cell line', annotation := line]
# metadata[annotation == "CD34-positive, CD38-positive common myeloid progenitor OR CD34-positive, CD38-positive common lymphoid progenitor", annotation := "CD34-positive, CD38-positive common progenitor"]
```


```{r}
annotation <- rtracklayer::import('suppa_analysis/gencode.v29.TSL12.gtf')
ioes <- rbindlist(lapply(to_analyze, function(event_type) fread(paste0(psi_input_dir, '/gencode.v29.TSL12.events_', event_type, '_strict.ioe'))))
ioes[, alternative_transcripts := lapply(strsplit(alternative_transcripts, ',', fixed = TRUE), I)]
ioes[, total_transcripts := lapply(strsplit(total_transcripts, ',', fixed = TRUE), I)]
ioes[, reference_transcripts := mapply(function(a, t) setdiff(t, a), alternative_transcripts, total_transcripts)]

# merge ioes into event_dt and add alternative_transcripts and total_transcripts to event_dt by reference
event_dt[ioes, on=.NATURAL, c('alternative_transcripts', 'reference_transcripts', 'total_transcripts') := .(alternative_transcripts, reference_transcripts, total_transcripts)]

event_transcripts <- event_dt[, .(transcript_id=unique(unlist(alternative_transcripts))), by=ID]
used_transcripts <- annotation[annotation$transcript_id %in% unique(event_transcripts$transcript_id) & annotation$type == 'transcript']
event_transcripts[data.table(transcript_id = used_transcripts$transcript_id, 
                             transcript_start = start(used_transcripts), 
                             transcript_end = end(used_transcripts)), on=.NATURAL, 
                  c('transcript_start', 'transcript_end') := .(transcript_start, transcript_end)]

other_transcripts <- event_dt[, .(transcript_id=unique(unlist(reference_transcripts))), by=ID]
used_other_transcripts <- annotation[annotation$transcript_id %in% unique(other_transcripts$transcript_id) & annotation$type == 'transcript']
other_transcripts[data.table(transcript_id = used_other_transcripts$transcript_id, 
                             transcript_start = start(used_other_transcripts), 
                             transcript_end = end(used_other_transcripts)), on=.NATURAL, 
                  c('transcript_start', 'transcript_end') := .(transcript_start, transcript_end)]

# used transcripts are the ones that include the alternative regions, i.e., that are part of the alternative transcripts
event_dt[event_transcripts[, .(alternative_start=list(transcript_start), alternative_end=list(transcript_end), gene_start=min(transcript_start), gene_end=max(transcript_end)), by=ID], on=.NATURAL, c('alternative_start', 'alternative_end', 'gene_start', 'gene_end') := .(alternative_start, alternative_end, gene_start, gene_end)]
event_dt[other_transcripts[, .(reference_start=list(transcript_start), reference_end=list(transcript_end)), by=ID], on=.NATURAL, c('reference_start', 'reference_end') := .(reference_start, reference_end)]

# used_genes <- annotation[annotation$gene_id %in% event_dt[, gene_id] & annotation$type == 'gene', ]
# event_dt[data.table(gene_id = used_genes$gene_id, gene_start = start(used_genes), gene_end = end(used_genes)), on = 'gene_id', c('gene_start', 'gene_end') := .(gene_start, gene_end)]

ss_dt <- merge(scores3, scores5, by = 'ID')
event_dt[ss_dt, on='ID', c('3ss', '5ss') := .(i.3ss, i.5ss)]

event_dt[, distance_TSS := ifelse(strand == '-', 
                                  log2(gene_end - end(event_gr[ID])), 
                                  log2(start(event_gr[ID]) - gene_start))]
# event_dt[strand == '-', distance_TSS:=log2(gene_end - end(event_gr[ID]))]
# event_dt[strand != '-', distance_TSS:=log2(start(event_gr[ID]) - gene_start)]

event_dt[, distance_TES := ifelse(strand == '-', 
                                  log2(start(event_gr[ID]) - gene_start), 
                                  log2(gene_end - end(event_gr[ID])))]
# event_dt[strand == '-', distance_TES:=log2(start(event_gr[ID]) - gene_start)]
# event_dt[strand != '-', distance_TES:=log2(gene_end - end(event_gr[ID]))]

# add stats about enhancers
# signal_psi_dt[, count_enhancer := as.integer(table(c(signal_psi_dt[, ID], from(enhancer_hits)))) - 1L]
    
# signal_psi_dt[data.table(ID = from(distances_enhancers), distance = mcols(distances_enhancers)$distance), on = .(ID), distance_enhancer := distance]

# load gene quantifications
# gene_quants <- readRDS('suppa_analysis/gene_expressions_TSL.rds')

# separator_gene_ids <- '_and_'
# signal_psi_dt <- event_dt[, .(ID, PSI = get(ihec), gene_id)]
# signal_psi_dt <- signal_psi_dt[, .(gene_id = unlist(tstrsplit(gene_id, separator_gene_ids, fixed=TRUE))), by=.(ID, PSI)]
# signal_psi_dt[unique(gene_quants[epirr_id_without_version == ihec, .(gene_id, gene_tpm)]), on = .(gene_id), gene_expression := gene_tpm]
# signal_psi_dt <- signal_psi_dt[, .(gene_id=paste(gene_id, collapse=separator_gene_ids), gene_expression=sum(gene_expression, na.rm = TRUE)), by=.(ID, PSI)]

    
# add region_lengths
event_dt[, paste('width', 'upstream_other_region', sep = ';') := log2(width(upstream_gr))]
event_dt[, paste('width', 'event_name', sep = ';') := log2(width(event_gr))]
event_dt[, paste('width', 'downstream_other_region', sep = ';') := log2(width(downstream_gr))]

# add distance to TSS and TES here
# signal_psi_dt[, distance_TSS := distance_TSS]
# signal_psi_dt[width(promoter_gr) == 1, distance_TSS := NA] # no distance for multi-gene-exons
# signal_psi_dt[, distance_TES := distance_TES]
# signal_psi_dt[width(tes_gr) == 1, distance_TES := NA] # no distance for multi-gene-exons
```



```{bash, eval=FALSE}
mkdir -p data
cd data 
# wget https://downloads.wenglab.org/Registry-V3/GRCh38-cCREs.bed
```

```{r}
# cCREs <- data.table::fread('data/GRCh38-cCREs.bed')
# names(cCREs) <- c('seqnames', 'start', 'end', 'some_id', 'accession', 'cCRE_type')
# cCRE_gr <- cCREs[, GRanges(seqnames = seqnames, IRanges(start = start, end = end), name = accession)]
# cCRE_hits <- findOverlaps(event_gr, cCRE_gr, maxgap = vicinity, ignore.strand=TRUE)
# cCREs_in_vicinity_ids <- sort(unique(to(cCRE_hits)[from(cCRE_hits) %in% keep_rows]))

activeChromHMM <- rtracklayer::import("StackedChromHMM_hg38_V1_EnhancerStates.bed.gz")
activeChromHMM$chrom_state <- activeChromHMM$name
activeChromHMM$name <- sprintf('chromhmm_%d', seq_along(activeChromHMM))
chromhmm_hits <- findOverlaps(event_gr, activeChromHMM, maxgap = vicinity, ignore.strand=TRUE)
chromhmm_in_vicinity <- sort(unique(to(chromhmm_hits)[from(chromhmm_hits) %in% keep_rows_manual]))

event_gr$name <- event_dt[, ID]
upstream_gr$name <- sprintf('up_%d', event_dt[, ID])
downstream_gr$name <- sprintf('down_%d', event_dt[, ID])

# write the bed files over which averaging should be done:
aggregateOver <- c(event_gr[keep_rows_manual],
                   upstream_gr[keep_rows_manual],
                   downstream_gr[keep_rows_manual],
                   # cCRE_gr[cCREs_in_vicinity_ids],
                   activeChromHMM[chromhmm_in_vicinity])
rtracklayer::export(aggregateOver, 'aggregateOver.bed')
```


```{r}
# aggregated_file_path <- file.path(psi_input_dir, 'aggregated_reference.rds')
save.image('aggregating.rda')
```

```{r}
# if (!file.exists(aggregated_file_path)) {
#   source('04-preliminary-results-aggregation.R') # or run with sbatch 04-preliminary-results-aggregation.sh
# }
# 
# aggregated_data <- readRDS(aggregated_file_path)

# at some point rename `Event Type` and other_region with ifelse(`Event Type` == 'SE', 'intron', 'exon')

# aggregated_data <- aggregated_data[sample_cols]
```
