---
title: "Correlation analysis"
output: pdf_document
---

```{r, fig.width=16, fig.height=10}
# if (!file.exists('aggregated_dt_filtered.rds'))
# saveRDS(aggregated_dt, 'aggregated_dt_filtered.rds')
load('aggregating.rda')
aggregated_dt <- fread('aggregated_dt_filtered.csv.gz', stringsAsFactors=TRUE)
# remove manual_ids, i.e., filter IDs in keep_rows
aggregated_dt <- aggregated_dt[ID %in% keep_rows]
# check if there are still manual_ids in the dt
stopifnot(!any(manual_ids %in% aggregated_dt$ID))
aggregated_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
```


```{r, fig.width=16, fig.height=10, eval=FALSE}
# vars_to_investigate <- c('0', '0.5', '<0.5')
# unlisted_vars <- unlist(unname(var_events), recursive = FALSE)
# unlisted_vars$'<0.5' <- setdiff(as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0'], use.names = FALSE)),
#                                 as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0.5'], use.names = FALSE)))
# unlisted_vars <- unlisted_vars[names(unlisted_vars) %in% vars_to_investigate]
```

```{r, eval=FALSE}
library(gprofiler2)
gostres <- gost(query = event_dt[as.integer(unlisted_vars$`0.5`), tstrsplit(as.character(unique(gene_id)), '.', fixed=TRUE, keep = 1)[[1]]],
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = FALSE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = .05, correction_method = "fdr", 
                domain_scope = "custom", custom_bg = event_dt[, tstrsplit(as.character(unique(gene_id)), '.', fixed=TRUE, keep = 1)[[1]]],
                numeric_ns = "", sources = c('GO'), as_short_link = FALSE)
gostplot(gostres, capped = TRUE, interactive = FALSE)
```


```{r, fig.width=16, fig.height=10}
# if cor_dt exists, read it, otherwise produce it:
if (file.exists('correlation.csv.gz')) {
  all_cor <- fread('correlation.csv.gz', stringsAsFactors = TRUE)
} else {
  cor_dt <-
    rbindlist(pbmclapply(setNames(cor_methods, cor_methods), function(cor_method)
      rbind(aggregated_dt[, lapply(.SD, function(y) {
        rmv <- is.na(y) | is.na(PSI)
        res <- NA
        if ((length(rmv) - sum(rmv)) >= minimum_events)
          res <-
          suppressWarnings(stats::cor(PSI, y, use = 'pairwise.complete.obs', method = cor_method))
        return(as.numeric(res))
      }), by = .(`Event Type`, IHEC, Variability), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type', 'gene_id', 'Variability')]],
      aggregated_dt[, lapply(.SD, function(y) {
        rmv <- is.na(y) | is.na(PSI)
        res <- NA
        if ((length(rmv) - sum(rmv)) >= minimum_events)
          res <-
          suppressWarnings(stats::cor(PSI, y, use = 'pairwise.complete.obs', method = cor_method))
        return(as.numeric(res))
      }), by = .(`Event Type`, IHEC), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type', 'gene_id', 'Variability')]][, Variability := 'All'])),
      idcol = 'Correlation Coefficient')
  # cor_dt[, `PSI Variance` := as.factor(`PSI Variance`)]
  cor_dt[, `Correlation Coefficient` := as.factor(`Correlation Coefficient`)]
  cor_dt[, `Partial Correlation` := FALSE]
  
  pcor_dt <-
    rbindlist(pbmclapply(setNames(cor_methods, cor_methods), function(cor_method)
      rbind(aggregated_dt[, lapply(.SD, function(y) {
        rmv <- is.na(y) | is.na(PSI) | is.na(gene_expression)
        res <- NA
        if ((length(rmv) - sum(rmv)) >= minimum_events)
          res <- tryCatch(
            suppressWarnings(
              ppcor::pcor.test(PSI[!rmv], y[!rmv], gene_expression[!rmv], method = cor_method)$estimate
            ),
            error = function(e) {
              if (identical(y, gene_expression) &
                  startsWith(e$message, prefix = 'system is computationally singular'))
                return(suppressWarnings(
                  stats::cor(y, PSI, use = 'na.or.complete', method = cor_method)
                ))
              return(e)
            }
          )
        return(as.numeric(res))
      }), by = .(`Event Type`, IHEC, Variability), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type', 'gene_id', 'Variability')]],
      aggregated_dt[, lapply(.SD, function(y) {
        rmv <- is.na(y) | is.na(PSI) | is.na(gene_expression)
        res <- NA
        if ((length(rmv) - sum(rmv)) >= minimum_events)
          res <- tryCatch(
            suppressWarnings(
              ppcor::pcor.test(PSI[!rmv], y[!rmv], gene_expression[!rmv], method = cor_method)$estimate
            ),
            error = function(e) {
              if (identical(y, gene_expression) &
                  startsWith(e$message, prefix = 'system is computationally singular'))
                return(suppressWarnings(
                  stats::cor(y, PSI, use = 'na.or.complete', method = cor_method)
                ))
              return(e)
            }
          )
        return(as.numeric(res))
      }), by = .(`Event Type`, IHEC), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type', 'gene_id', 'Variability')]][, Variability := 'All'])),
      idcol = 'Correlation Coefficient')
  # pcor_dt[, `PSI Variance` := as.factor(`PSI Variance`)]
  pcor_dt[, `Correlation Coefficient` := as.factor(`Correlation Coefficient`)]
  pcor_dt[, `Partial Correlation` := TRUE]
  
  
  all_cor <- rbind(cor_dt, pcor_dt)
  all_cor[, Variability := factor(Variability,
                                  levels = c('Low Variability', 'All', 'High Variability'))]
  # all_cor[`PSI Variance` == '0', `PSI Variance`:='all']
  # all_cor[`PSI Variance` == '0.5', `PSI Variance`:='high']
  # all_cor[`PSI Variance` == '<0.5', `PSI Variance`:='low']
  # all_cor[, `PSI Variance`:=factor(`PSI Variance`, levels = c('low', 'all', 'high'))]
  fwrite(all_cor, 'correlation.csv.gz')
}
```


```{r, fig.width=16, fig.height=10, eval=FALSE}
# for (this_event in to_analyze) {
#   var_events_type <- var_events[[this_event]]
#   for (var_threshold in names(var_events_type)) {
#     for (this_cor_method in cor_methods) {
#       for (this_partial in c(TRUE, FALSE)) {
#         cor_dt_event_var <-
#           all_cor[`Event Type` == this_event &
#                     `PSI Variance` == var_threshold &
#                     `Correlation Coefficient` == this_cor_method &
#                     `Partial Correlation` == this_partial]
#         cor_mat <-
#           as.matrix(cor_dt_event_var[,-c('IHEC',
#                                          'Event Type',
#                                          'PSI Variance',
#                                          'Partial Correlation',
#                                          'Correlation Coefficient')])
#         rownames(cor_mat) <- cor_dt_event_var[, IHEC]
#         breaks <- seq(-.4, .4, by = .05)
#         annotation_df <-
#           metadata[epirr_id_without_version %in% sample_cols, data.frame(annotation = annotation, row.names = epirr_id_without_version)]
#         tryCatch({
#         #   pheatmap(
#         #     t(cor_mat[sample_cols,])[, rownames(annotation_df)[order(annotation_df$annotation)]],
#         #     color = colorRampPalette(c("red", "white", "blue"))(length(breaks)),
#         #     breaks = breaks,
#         #     annotation_col = annotation_df,
#         #     cluster_cols = FALSE,
#         #     cluster_rows = FALSE,
#         #     main = paste(
#         #       this_event,
#         #       ':',
#         #       this_cor_method,
#         #       ', PSI Variance Quantile:',
#         #       var_threshold,
#         #       ', partial',
#         #       this_partial
#         #     ),
#         #     na_col = 'grey',
#         #     show_colnames = FALSE
#         # )
#         }, error = function(e) {
#           print(paste(this_event, var_threshold, e$message))
#         })
#       }
#     }
#   }
# }
```

```{r, fig.width=12, fig.height=8}
cor_dt_intermediate <- copy(all_cor)
# below_0.5 <- aggregated_dt[ID %in% setdiff(as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0'], use.names = FALSE)),
#         as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0.5'], use.names = FALSE))), lapply(.SD, function(y) {
#         rmv <- is.na(y) | is.na(PSI) | is.na(gene_expression)
#         res <- NA
#         if ((length(rmv) - sum(rmv)) >= minimum_events)
#           res <- tryCatch(
#             suppressWarnings(ppcor::pcor.test(PSI[!rmv], y[!rmv], gene_expression[!rmv], method = 'pearson')$estimate),
#             error = function(e) {
#               if (identical(y, gene_expression) &
#                   startsWith(e$message, prefix = 'system is computationally singular'))
#                 return(suppressWarnings(stats::cor(y, PSI, use = 'pairwise', method = 'pearson')))
#               return(e)
#             }
#           )
#         return(as.numeric(res))
#       }), by = .(`Event Type`, IHEC), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type')]]
# below_0.5[, c('PSI Variance', 'Partial Correlation', 'Correlation Coefficient'):=list('<0.5', TRUE, 'pearson')]
# cor_dt_intermediate <- rbind(cor_dt_intermediate, below_0.5)
# colnames(cor_dt_intermediate) <- gsub(';max', '', colnames(cor_dt_intermediate), fixed = TRUE)
# colnames(cor_dt_intermediate) <- gsub(';mean', '', colnames(cor_dt_intermediate), fixed = TRUE)
melt_cor_dt <- melt(cor_dt_intermediate, id.vars = c('IHEC', 'Event Type', 'Variability', 'Partial Correlation', 'Correlation Coefficient'), variable.name = 'feature', value.name = 'Correlation w/ PSI')
melt_cor_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
melt_cor_dt[`Event Type` == 'SE', feature := gsub('other_region', 'Intron', feature, fixed=TRUE)]
melt_cor_dt[`Event Type` == 'RI', feature := gsub('other_region', 'Exon', feature, fixed=TRUE)]
melt_cor_dt[`Event Type` == 'SE', feature := gsub('event_name', 'SE', feature, fixed=TRUE)]
melt_cor_dt[`Event Type` == 'RI', feature := gsub('event_name', 'RI', feature, fixed=TRUE)]
# melt_cor_dt[, feature:=sub('wgbs', 'DNAm', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub(';', '\n', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub(',', '\n', feature, fixed = TRUE)]
melt_cor_dt[, feature:=gsub('_', ' ', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('distance ', 'Distance\n', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('upstream Intron', 'Intron\nUpstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('upstream Exon', 'Exon\nUpstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('downstream Intron', 'Intron\nDownstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('downstream Exon', 'Exon\nDownstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('gene expression', 'Gene\nExpression', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('width', 'Width', feature, fixed = TRUE)]
aggregate_median_by <- c('feature', 'Event Type', 'Variability', 'Partial Correlation', 'Correlation Coefficient')
melt_cor_dt[, median_cor:=median(`Correlation w/ PSI`, na.rm = TRUE), by = aggregate_median_by]
melt_cor_dt[, mean_cor:=mean(`Correlation w/ PSI`, na.rm = TRUE), by = aggregate_median_by]
melt_cor_dt[, sd_cor:=sd(`Correlation w/ PSI`, na.rm = TRUE), by = aggregate_median_by]
melt_cor_dt[, relevant:=any(abs(median_cor) > .1), by = aggregate_median_by]
melt_cor_dt[, Variability:=sub(' Variability', '', Variability, fixed = TRUE)]
melt_cor_dt[, Variability:=factor(Variability, levels = c('Low', 'All', 'High'))]
```

```{r, fig.width=9, fig.height=5}
 # "#000000"     "#E69F00"     "#56B4E9"     "#009E73"     "#F0E442"     "#0072B2"     "#D55E00"     "#CC79A7"     "#999999" 
ihec_ia_colors <- unlist(jsonlite::read_json('IHEC_EpiATLAS_IA_colors_Apr_2024_Final.json'), recursive = FALSE)
hex_colors <- sapply(unlist(ihec_ia_colors$experiment, recursive = FALSE), function(x) {
    cols <- as.numeric(strsplit(x, ",")[[1]])
    rgb(cols[1], cols[2], cols[3], maxColorValue = 255)
})
hex_colors <- c(hex_colors, DNAm=hex_colors[['WGBS']])

top_n <- 10

tmp_cor_dt <- melt_cor_dt[`Variability` == 'All' & `Correlation Coefficient` == 'pearson' & `Partial Correlation` == 'TRUE']
tmp_cor_dt[order(-abs(median_cor)), .(feature=unique(feature)[1:top_n]), by=.(`Event Type`)]
tmp_cor_dt[, 'Feature Type' := ifelse(Reduce(`|`, lapply(c(histone_marks, 'DNAm'), function(x) startsWith(feature, x))), 'Epigenetic', 'Intrinsic')]
tmp_cor_dt[feature %in% c('max promoter', 'summed enhancer'), `Feature Type` := 'Epigenetic']
plot_dt <- merge(tmp_cor_dt, tmp_cor_dt[order(-abs(median_cor)), .(feature=unique(feature)[1:top_n]), by=.(`Event Type`)])
# Create a vector of colors for ggplot, defaulting to white
plot_colors <- setNames(rep("dark grey", plot_dt[, uniqueN(feature)]), plot_dt[, unique(feature)])
for(name in names(hex_colors)) {
  inds <- startsWith(names(plot_colors), name)
  plot_colors[inds] <- hex_colors[name]
}

ggplot(plot_dt, 
       aes(y = `Correlation w/ PSI`, 
           x =  tidytext::reorder_within(feature, -median_cor, `Event Type`),
       color = feature)) + #`Feature Type`)) + 
    geom_hline(yintercept = 0, color = "light grey", linetype = "dashed") + 
    geom_boxplot() + 
    labs(x = NULL, #'Aggregated Signal or Statistic', 
         title = NULL#, paste('Top', top_n, 'Globally Correlated Features')
         ) +
    tidytext::scale_x_reordered() + 
    facet_wrap(. ~ `Event Type`, scales="free", nrow=2) + 
    scale_color_manual(values = plot_colors) +
    theme_bw(base_size=12) + theme(legend.position = 'none', strip.background = element_rect(fill = 'white'))
ggsave(filename = file.path(plot_dir, paste0('top', top_n, '_overall.png')), width = 10, height = 5)
ggsave(filename = file.path(plot_dir, paste0('top', top_n, '_overall.svg')), width = 10, height = 5)
```


```{r, fig.width=9, fig.height=5}
for (color_by in c('Correlation Coefficient', 'Partial Correlation', 'Variability')) {
  find_relevant_by <- aggregate_median_by[aggregate_median_by != color_by]
  melt_cor_dt[, plot_relevant:=NULL]
  melt_cor_dt[, plot_relevant:=any(relevant), by = mget(find_relevant_by)]
  breaks <- seq(-.5, .5, by = .1)
  not_color_by <- TRUE
  if (color_by == 'Partial Correlation') {
    not_color_by <- melt_cor_dt[, `Correlation Coefficient` == 'pearson' & `Variability` == 'All']
  } else if (color_by == 'Correlation Coefficient') {
    not_color_by <- melt_cor_dt[, `Partial Correlation` == 'TRUE'  & `Variability` == 'All']
  } else if (color_by == 'Variability') {
    not_color_by <- melt_cor_dt[, `Partial Correlation` == 'TRUE' & `Correlation Coefficient` == 'pearson']
  }
  
  color_levels <- melt_cor_dt[not_color_by, uniqueN(get(color_by))]
  tmp_cor_dt <- melt_cor_dt[not_color_by & !is.na(`Correlation w/ PSI`), if(uniqueN(get(color_by)) == color_levels) .SD, by=.(`Event Type`, feature)]
  if (color_by == 'Variability') {
    median_dt <- dcast(tmp_cor_dt[, .(median_cor=unique(median_cor)), by=.(`Event Type`, feature, `Variability`)], `Event Type` + feature ~ `Variability`, value.var = 'median_cor')
    median_dt[, diff_col:=abs(`Low`-`High`)]
    sorted_idx <- median_dt[, order(-diff_col)]
    p <- ggplot(merge(tmp_cor_dt, median_dt[sorted_idx, .(feature=feature[1:top_n], diff_col=diff_col[1:top_n], all_median=All[1:top_n]), by=.(`Event Type`)]), 
       aes(y = `Correlation w/ PSI`, 
           x =  tidytext::reorder_within(feature, -diff_col, `Event Type`), 
           fill = get(color_by))) + 
    theme_bw() + geom_hline(yintercept = 0, color = "darkgrey", linetype = "dashed") + 
    geom_boxplot() + 
    labs(x = 'Aggregated Signal or Statistic') +
    tidytext::scale_x_reordered() + 
    facet_wrap(. ~ `Event Type`, scales="free", nrow=2) + 
    theme(legend.position = 'bottom') +
    scale_fill_manual(values = variability_colors[1:color_levels], name = color_by)
    ggsave(filename = file.path(plot_dir, paste0('max diff ', color_by, '.pdf')), p, width = 10, height = 7) 
    print(p)
  }
p <- ggplot(merge(tmp_cor_dt, tmp_cor_dt[order(-abs(median_cor)), .(feature=unique(feature)[1:top_n]), by=.(`Event Type`)]), 
             aes(y = `Correlation w/ PSI`, 
                 x =  tidytext::reorder_within(feature, -median_cor, `Event Type`), 
                 color = get(color_by))) + 
        theme_bw() + geom_hline(yintercept = 0, color = "darkgrey", linetype = "dashed") + 
    geom_boxplot() + 
    labs(x = 'Aggregated Signal or Statistic') +
    tidytext::scale_x_reordered() + 
    facet_wrap(. ~ `Event Type`, scales="free", nrow=2) +
    theme(legend.position = 'bottom', strip.background = element_rect(fill = 'white')) +
    scale_color_manual(values = variability_colors, name = color_by)
if (color_levels == 2){
  p <- p + scale_color_manual(values = unname(variability_colors[c(1,3)]), name = color_by)}
ggsave(filename = file.path(plot_dir, paste0('top', top_n,  color_by, '.pdf')), p, width = 10, height = 6)
if (color_by == "Variability")
ggsave(filename = file.path(publication_plots_dir, paste0('top', top_n,  color_by, '.pdf')), p, width = 7, height = 4.5)
  print(p)
}
```
```{r}
heatmap_dt <- melt_cor_dt[`Partial Correlation` == TRUE & `Correlation Coefficient` == 'pearson' & `Variability` == 'All' & grepl(paste(c(histone_marks, "DNAm"), collapse = "|"), x = feature)]
heatmap_dt[, mark:=tstrsplit(x = feature, split = "\n", fixed = TRUE, keep = 1)]
heatmap_dt[, region:=sub(pattern = ".*?\n", replacement = "", x = feature)]
ggplot(heatmap_dt, aes(x = factor(region, levels = c("Intron\nUpstream", "SE", "Intron\nDownstream", "Exon\nUpstream", "RI", "Exon\nDownstream")), 
                                                                                                                              y = factor(mark, levels = rev(c("H3K9me3", "H3K27me3", "H3K4me3", "H3K4me1", "H3K27ac", "H3K36me3", "DNAm"))), 
                                                                                                                              fill = mean_cor)) + 
    geom_tile() + 
    ggh4x::facet_grid2( ~ `Event Type`, scales = 'free') +
    theme_bw() + 
    theme(strip.background = element_rect(fill = 'white'), legend.position = "bottom") + 
    scale_fill_gradient2(
  low = scales::muted("blue"),
  mid = "white",
  high = scales::muted("red")) + 
  # add text into the tile with the value of the mean and sd_cor in parantheses
  geom_text(aes(label = paste0(round(mean_cor, 2), " (\u00B1", round(sd_cor, 2), ")")), size=3) +
  
    labs(y = 'Mark', x = 'Region', fill = 'Partial Correlation')
ggsave(filename = file.path(plot_dir, "epigenetic_cor.pdf"), width = 7, height = 4.5)
ggsave(filename = file.path(publication_plots_dir, "epigenetic_cor.pdf"), width = 7, height = 4.5)
```

```{r, eval=FALSE}
ggplot(melt_cor_dt[feature == 'DNAm\nPLS' & `Partial Correlation` == 'TRUE' & `Correlation Coefficient` == 'pearson'], aes(y=`Correlation w/ PSI`, x=`Variability`)) + geom_dotplot(binaxis='y', stackdir='center', dotsize=.2) + geom_boxplot(width=0.1) + theme_bw() + facet_wrap(. ~ `Event Type`, nrow = 2)
```

```{r, fig.width=10, fig.height=10}
stopifnot(nrow(dcast(all_cor, formula = `Variability` + `Correlation Coefficient` + IHEC + `Event Type` ~ `Partial Correlation`, value.var = 'gene_expression')[`FALSE` - `TRUE` > 1e-10]) == 0)
# test_dt <- aggregated_dt[`Event Type` == 'SE' & ID %in% event_dt[!is.na(`PSI SD`), ID] & IHEC == sample(IHEC, 1)]
# stopifnot(test_dt[, cor(PSI, gene_expression, use = 'pairwise', 'pearson')] !=
# test_dt[!(is.na(PSI) | is.na(gene_expression)), ppcor::pcor.test(PSI, gene_expression, gene_expression, 'pearson')]$estimate)
# test_dt[, cor(PSI, gene_expression, use = 'pairwise', 'pearson')]
# test_dt[!(is.na(PSI) | is.na(gene_expression)), ppcor::pcor.test(PSI, gene_expression, gene_expression, 'pearson')]$estimate
# test_dt[!(is.na(PSI) | is.na(gene_expression)), ppcor::pcor.test(gene_expression, PSI, gene_expression, 'pearson')]$estimate
```

# By ID:
```{r, fig.width=16, fig.height=10, eval=FALSE}
cor_dt <-
  rbindlist(setNames(pbmclapply(unique(
    names(unlisted_vars)
  ), function(var) {
    print(var)
    var_ids <-
      as.integer(unlist(unlisted_vars[names(unlisted_vars) == var], use.names = FALSE))
    rbindlist(sapply(cor_methods, function(cor_method)
      aggregated_dt[ID %in% var_ids, lapply(.SD, function(y) {
        rmv <- is.na(y) | is.na(PSI)
        res <- NA
        if ((length(rmv) - sum(rmv)) >= minimum_events)
          res <-
          suppressWarnings(stats::cor(PSI, y, use = 'na.or.complete', method = cor_method))
        return(as.numeric(res))
      }), by = .(`Event Type`, ID), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type')]]
      , simplify = FALSE),
      idcol = 'Correlation Coefficient')
  }), unique(names(unlisted_vars))), idcol = 'PSI Variance')
cor_dt[, `PSI Variance` := as.factor(`PSI Variance`)]
cor_dt[, `Correlation Coefficient` := as.factor(`Correlation Coefficient`)]
cor_dt[, `Partial Correlation` := FALSE]


pcor_dt <-
  rbindlist(setNames(pbmclapply(unique(
    names(unlisted_vars)
  ), function(var) {
    print(var)
    var_ids <-
      as.integer(unlist(unlisted_vars[names(unlisted_vars) == var], use.names = FALSE))
  rbindlist(sapply(cor_methods, function(cor_method)
      aggregated_dt[ID %in% var_ids, lapply(.SD, function(y) {
        rmv <- is.na(y) | is.na(PSI) | is.na(gene_expression)
        res <- NA
        if ((length(rmv) - sum(rmv)) >= minimum_events)
          res <- tryCatch(
            suppressWarnings(ppcor::pcor.test(PSI[!rmv], y[!rmv], gene_expression[!rmv], method = cor_method)$estimate),
            error = function(e) {
              if (identical(y, gene_expression) &
                  startsWith(e$message, prefix = 'system is computationally singular'))
                return(suppressWarnings(stats::cor(y, PSI, use = 'na.or.complete', method = cor_method)))
              return(e)
            }
          )
        return(as.numeric(res))
      }), by = .(`Event Type`, ID), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type')]]
      , simplify = FALSE),
      idcol = 'Correlation Coefficient')
  }), unique(names(unlisted_vars))), idcol = 'PSI Variance')
pcor_dt[, `PSI Variance` := as.factor(`PSI Variance`)]
pcor_dt[, `Correlation Coefficient` := as.factor(`Correlation Coefficient`)]
pcor_dt[, `Partial Correlation` := TRUE]


all_cor <- rbind(cor_dt, pcor_dt)
all_cor[`PSI Variance` == '0', `PSI Variance`:='all']
all_cor[`PSI Variance` == '0.5', `PSI Variance`:='high']
all_cor[`PSI Variance` == '<0.5', `PSI Variance`:='low']
all_cor[, `PSI Variance`:=factor(`PSI Variance`, levels = c('low', 'all', 'high'))]
```

```{r, fig.width=12, fig.height=8, eval=FALSE}
cor_dt_intermediate <- copy(all_cor)
# below_0.5 <- aggregated_dt[ID %in% setdiff(as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0'], use.names = FALSE)),
#         as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0.5'], use.names = FALSE))), lapply(.SD, function(y) {
#         rmv <- is.na(y) | is.na(PSI) | is.na(gene_expression)
#         res <- NA
#         if ((length(rmv) - sum(rmv)) >= minimum_events)
#           res <- tryCatch(
#             suppressWarnings(ppcor::pcor.test(PSI[!rmv], y[!rmv], gene_expression[!rmv], method = 'pearson')$estimate),
#             error = function(e) {
#               if (identical(y, gene_expression) &
#                   startsWith(e$message, prefix = 'system is computationally singular'))
#                 return(suppressWarnings(stats::cor(y, PSI, use = 'pairwise', method = 'pearson')))
#               return(e)
#             }
#           )
#         return(as.numeric(res))
#       }), by = .(`Event Type`, IHEC), .SDcols = names(aggregated_dt)[!names(aggregated_dt) %in% c('ID', 'PSI', 'IHEC', 'Event Type')]]
# below_0.5[, c('PSI Variance', 'Partial Correlation', 'Correlation Coefficient'):=list('<0.5', TRUE, 'pearson')]
# cor_dt_intermediate <- rbind(cor_dt_intermediate, below_0.5)
# colnames(cor_dt_intermediate) <- gsub(';max', '', colnames(cor_dt_intermediate), fixed = TRUE)
# colnames(cor_dt_intermediate) <- gsub(';mean', '', colnames(cor_dt_intermediate), fixed = TRUE)
melt_cor_dt <- melt(cor_dt_intermediate, id.vars = c('ID', 'Event Type', 'PSI Variance', 'Partial Correlation', 'Correlation Coefficient'), variable.name = 'feature', value.name = 'Correlation w/ PSI')
melt_cor_dt[`Event Type` == 'SE', feature := gsub('other_region', 'Intron', feature, fixed=TRUE)]
melt_cor_dt[`Event Type` == 'RI', feature := gsub('other_region', 'Exon', feature, fixed=TRUE)]
melt_cor_dt[`Event Type` == 'SE', feature := gsub('event_name', 'SE', feature, fixed=TRUE)]
melt_cor_dt[`Event Type` == 'RI', feature := gsub('event_name', 'RI', feature, fixed=TRUE)]
# melt_cor_dt[, feature:=sub('wgbs', 'DNAm', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub(';', '\n', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub(',', '\n', feature, fixed = TRUE)]
melt_cor_dt[, feature:=gsub('_', ' ', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('distance ', 'Distance\n', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('upstream Intron', 'Intron\nUpstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('upstream Exon', 'Exon\nUpstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('downstream Intron', 'Intron\nDownstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('downstream Exon', 'Exon\nDownstream', feature, fixed = TRUE)]
melt_cor_dt[, feature:=sub('width', 'Width', feature, fixed = TRUE)]
aggregate_median_by <- c('feature', 'Event Type', 'PSI Variance', 'Partial Correlation', 'Correlation Coefficient')
melt_cor_dt[, median_cor:=median(`Correlation w/ PSI`, na.rm = TRUE), by = aggregate_median_by]
melt_cor_dt[, relevant:=any(abs(median_cor) > .1), by = aggregate_median_by]
```


```{r, fig.width=8, fig.height=8, eval=FALSE}
my_colors <- c("#CC79A7", "#999999", "#009E73")

tmp_cor_dt <- melt_cor_dt[`PSI Variance` == 'all' & `Correlation Coefficient` == 'pearson' & `Partial Correlation` == 'TRUE']
tmp_cor_dt[order(-abs(median_cor)), .(feature=unique(feature)[1:10]), by=.(`Event Type`)]
ggplot(merge(tmp_cor_dt, tmp_cor_dt[order(-abs(median_cor)), .(feature=unique(feature)[1:10]), by=.(`Event Type`)]), 
       aes(y = `Correlation w/ PSI`, 
           x =  tidytext::reorder_within(feature, -median_cor, `Event Type`),
       fill = '#999999')) + 
    theme_bw() + geom_hline(yintercept = 0, color = "darkgrey", linetype = "dashed") + 
    geom_violin() + 
    labs(x = 'Aggregated Signal or Statistic') +
    tidytext::scale_x_reordered() + 
    facet_wrap(. ~ `Event Type`, scales="free", nrow=2) +
    scale_fill_manual(values = '#999999') +
    theme(legend.position = 'none')
ggsave(filename = file.path(plot_dir, 'top10_overall_event.pdf'), width = 9, height = 5)
```


```{r, fig.width=9, fig.height=5, eval=FALSE}
for (color_by in c('Correlation Coefficient', 'Partial Correlation', 'PSI Variance')) {
  find_relevant_by <- aggregate_median_by[aggregate_median_by != color_by]
  melt_cor_dt[, plot_relevant:=NULL]
  melt_cor_dt[, plot_relevant:=any(relevant), by = mget(find_relevant_by)]
  breaks <- seq(-.5, .5, by = .1)
  not_color_by <- TRUE
  if (color_by == 'Partial Correlation') {
    not_color_by <- melt_cor_dt[, `Correlation Coefficient` == 'pearson' & `PSI Variance` == 'all']
  } else if (color_by == 'Correlation Coefficient') {
    not_color_by <- melt_cor_dt[, `Partial Correlation` == 'TRUE'  & `PSI Variance` == 'all']
  } else if (color_by == 'PSI Variance') {
    not_color_by <- melt_cor_dt[, `Partial Correlation` == 'TRUE' & `Correlation Coefficient` == 'pearson']
  }
  
  color_levels <- melt_cor_dt[not_color_by, uniqueN(get(color_by))]
  tmp_cor_dt <- melt_cor_dt[not_color_by & !is.na(`Correlation w/ PSI`), if(uniqueN(get(color_by)) == color_levels) .SD, by=.(`Event Type`, feature)]
  if (color_by == 'PSI Variance') {
    median_dt <- dcast(tmp_cor_dt[, .(median_cor=unique(median_cor)), by=.(`Event Type`, feature, `PSI Variance`)], `Event Type` + feature ~ `PSI Variance`, value.var = 'median_cor')
    median_dt[, diff_col:=abs(low-high)]
    sorted_idx <- median_dt[, order(-diff_col)]
    p <- ggplot(merge(tmp_cor_dt, median_dt[sorted_idx, .(feature=feature[1:10], diff_col=diff_col[1:10], all_median=all[1:10]), by=.(`Event Type`)]), 
       aes(y = `Correlation w/ PSI`, 
           x =  tidytext::reorder_within(feature, -all_median, `Event Type`), 
           fill = get(color_by))) + 
    theme_bw() + geom_hline(yintercept = 0, color = "darkgrey", linetype = "dashed") + 
    geom_violin() + 
    labs(x = 'Aggregated Signal or Statistic') +
    tidytext::scale_x_reordered() + 
    facet_wrap(. ~ `Event Type`, scales="free", nrow=2) + 
    scale_fill_manual(values = my_colors[1:color_levels], name = color_by)
    ggsave(filename = file.path(plot_dir, paste0('max diff ', color_by, '_event.pdf')), p, width = 10, height = 7)
    print(p)
  }
p <- ggplot(merge(tmp_cor_dt, tmp_cor_dt[order(-abs(median_cor)), .(feature=unique(feature)[1:10]), by=.(`Event Type`)]), 
             aes(y = `Correlation w/ PSI`, 
                 x =  tidytext::reorder_within(feature, -median_cor, `Event Type`), 
                 fill = get(color_by))) + 
        theme_bw() + geom_hline(yintercept = 0, color = "darkgrey", linetype = "dashed") + 
    geom_violin() + 
    labs(x = 'Aggregated Signal or Statistic') +
    tidytext::scale_x_reordered() + 
    facet_wrap(. ~ `Event Type`, scales="free", nrow=2) +
    scale_fill_manual(values = my_colors, name = color_by)
if (color_levels == 2)
  p <- p + scale_fill_manual(values = my_colors[c(1,3)], name = color_by)
ggsave(filename = file.path(plot_dir, paste0('top10 ', color_by, '_event.pdf')), p, width = 10, height = 7)
  print(p)
}
```


```{r, fig.width=10, fig.height=10, eval=FALSE}
p <- ggplot(aggregated_dt[`Event Type` == 'SE' & ID %in% as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0'], use.names = FALSE)) & IHEC == 'IHECRE00000060'], aes(x = `width;event_name`, y = PSI, color = `Event Type`)) + facet_wrap(. ~ IHEC) +  geom_point() +
  ggpubr::stat_cor(colour = "black")+ ggpubr::stat_cor(method='spearman', cor.coef.name = 'rho', label.y.npc = 'bottom', colour = "black")
p1 <- ggplot(aggregated_dt[`Event Type` == 'SE' & ID %in% as.integer(unlist(unlisted_vars[names(unlisted_vars) == '0'], use.names = FALSE)) & IHEC == 'IHECRE00000060'], aes(x = `width;event_name`, y = PSI, color = `Event Type`)) + facet_wrap(. ~ IHEC) +  geom_point() + scale_x_log10() + 
  ggpubr::stat_cor(colour = "black")+ ggpubr::stat_cor(method='spearman', cor.coef.name = 'rho', label.y.npc = 'bottom', colour = "black")
p + p1
```
