---
title: "Post-aggregation analysis"
output: html_document
---

```{r, load-data}
# load the image
load('processed_data/aggregating.rda')
# load the files table
file_table <- fread('processed_data/file_table.csv.gz')
# we only need ChIP-Seq data, since WGBS has been aggregated via the matrices
file_table <- file_table[assay_type == "ChIP-Seq"]
# locate
file_table[, local_file := file.path(sample_dt_dir, paste0(basename(file_path), '.tab.gz'))]
stopifnot(file_table[, all(file.exists(local_file))])

setkey(file_table, 'local_file')
# write table back to file
fwrite(file_table, 'processed_data/file_table.csv.gz')

# load WGBS data
wgbs_agg <- fread(file.path(sample_dt_dir, 'WGBS_agg.csv.gz'))
wgbs_agg <- wgbs_agg[!startsWith(name, 'chromhmm')]
wgbs_agg[, experiment_type:='DNAm']
wgbs_agg[, ihec := sub("\\.[0-9]+$", "", ihec)]

# bind all the samples for the histone marks
histone_dt <- rbindlist(
  pbmcapply::pbmcmapply(
    FUN = function(file) {
      dt <-
        fread(file)
      names(dt) <-
        c('name', 'size', 'covered', 'sum', 'mean0', 'mean', 'min', 'max')
      dt[!startsWith(name, 'chromhmm')]
    },
    file = file_table[, local_file],
    USE.NAMES = TRUE,
    SIMPLIFY = FALSE
  ),
  idcol = 'local_file'
)
# some mapping for ids and histone marks
histone_dt[, ihec:=file_table[local_file, epirr_id_without_version]]
histone_dt[, experiment_type:=file_table[local_file, experiment_type]]
setnames(histone_dt, old = 'mean', new = 'score')

# find common cols and rows
common_ihec <- intersect(histone_dt[, unique(ihec)], wgbs_agg[, unique(ihec)])
common_cols <- intersect(names(histone_dt), names(wgbs_agg))
```


```{r, build-aggregated-dt}
# bind histone and WGBS data
aggregated_dt <- rbindlist(list(histone_dt[ihec %in% common_ihec, ..common_cols], 
                                wgbs_agg[ihec %in% common_ihec, ..common_cols]))
# some data wrangling to prepare the data for putting into wide format 
aggregated_dt[, c("position", "ID"):=tstrsplit(name, "_", fixed=TRUE)]
aggregated_dt[is.na(ID), c("position", "ID"):=list('event_name', position)]
aggregated_dt[position=="up", position:="upstream_other_region"]
aggregated_dt[position=="down", position:="downstream_other_region"]
aggregated_dt[, ihec:=as.factor(ihec)]
aggregated_dt[, ID:=as.integer(ID)]
# from long to wide
aggregated_dt <- dcast(aggregated_dt, ihec + ID ~ experiment_type + position, value.var = "score", sep = ";")
setnames(aggregated_dt, old = 'ihec', new = 'IHEC')
# write full table to disk
fwrite(aggregated_dt, 'processed_data/aggregated_dt.csv.gz')
```


```{r, melt-psi}
# bring PSI values to long format for compatibility with the aggregated table
melted_psi <- melt(event_dt, measure.vars = sample_cols, id.vars = c('ID', 'Event Type', 'gene_id'), variable.name = 'IHEC', value.name = 'PSI', variable.factor = FALSE)
melted_psi <- melted_psi[!is.na(PSI)]
```


```{r, expand-aggregated-dt}
# add PSI to the aggregated table
aggregated_dt[melted_psi[ID %in% keep_rows_manual & !is.na(PSI)], on=.(ID, IHEC), PSI:=PSI]
# remove NA PSIs
aggregated_dt <- aggregated_dt[!is.na(PSI)]

# gather the intrinsic features
cols_to_add <- c('gene_id', 'distance_TSS', 'distance_TES', 'width;upstream_other_region', 'width;event_name', 'width;downstream_other_region', 'Event Type', '3ss', '5ss', 'Variability')
# add the intrinsic features
aggregated_dt[event_dt, on=.(ID), (cols_to_add):=mget(cols_to_add)]

# read the gene expression
gene_quants <- readRDS('suppa_analysis/gene_expressions.rds')

# manually change gene_id for ENSG00000213593.9_and_ENSG00000254462.1_and_ENSG00000211450.9_and_ENSG00000198561.13 from the manual ids that were kept
aggregated_dt[gene_id == "ENSG00000213593.9_and_ENSG00000254462.1_and_ENSG00000211450.9_and_ENSG00000198561.13", gene_id := "ENSG00000198561.13"]
aggregated_dt[gene_quants, on=c(gene_id='gene_id', IHEC='epirr_id_without_version'), gene_expression:=gene_tpm]

# make sure all of the gene_expression values are at least one:
stopifnot(all(aggregated_dt[, gene_expression >= 1]))
# log the expression
aggregated_dt[, gene_expression:=log2(gene_expression)]

# remove the gene_id version
aggregated_dt[, gene_id:=gsub(pattern = '.\\d+$', '', gene_id)]
# get the epirrs for which there is observed H3K27ac because only they have max_promoter and summed_enhancer
ihec_of_interest <- file_table[experiment_type == 'H3K27ac' & uuid != "imputed", .(IHEC=epirr_id_without_version, uuid)]

# read the max promoter
max_promoter <- fread('data/GeneMaxPromoterSignal.txt.gz') # from /IHEC/post_processed_data/Analysis/IHEC_gABC_Interactions/GeneMaxPromoterSignal.txt.gz
promoter_melt <- melt(max_promoter, id.vars = 'Ensembl ID', variable.name = 'uuid', value.name = 'max_promoter', variable.factor = FALSE)
promoter_merged <- promoter_melt[ihec_of_interest, on=.(uuid)][, .(IHEC, uuid, gene_id=`Ensembl ID`, max_promoter)]

# read the summed enhancer
summed_enhancer <- fread('data/gABC_GeneSummedEnhancer.txt.gz') # from /IHEC/post_processed_data/Analysis/IHEC_gABC_Interactions/gABC_GeneSummedEnhancer.txt.gz
enhancer_melt <- melt(summed_enhancer, id.vars = 'Ensembl ID', variable.name = 'uuid', value.name = 'summed_enhancer', variable.factor = FALSE)
promoter_merged[enhancer_melt, on=c(uuid='uuid', gene_id='Ensembl ID'), summed_enhancer:=summed_enhancer]

# add the max promoter and summed enhancer to the aggregated table
aggregated_dt[promoter_merged, on=.(IHEC, gene_id), c('summed_enhancer', 'max_promoter'):=.(summed_enhancer, max_promoter)]
```


```{r, write-filtered}
# write the filtered and augmented table to file
fwrite(aggregated_dt, 'processed_data/aggregated_dt_filtered.csv.gz')
```
