---
title: "ML"
output: pdf_document
---

# overall model

```{r, eval=FALSE}
aggregated_dt <- readRDS('aggregated_dt.rds')
aggregation <- 'max'
response <- 'PSI'
```


```{r, eval=FALSE}
if (!all(file.exists(paste(to_analyze, 'models.rds', sep = '_'))))
source('06-ml.R')
```

# check models

```{r, fig.width=16, fig.height=11, eval=FALSE}
rank_dt <- rbindlist(sapply(to_analyze, function(this_event){
  models <- all_models[, this_event]
  as_dt <- rbindlist(sapply(names(models), function(cvfit_name) {
    cvfits <- models[[cvfit_name]]$cvfit
    rbindlist(sapply(names(cvfits), function(alpha) {
      cvfit <- cvfits[[alpha]]
      if (endsWith(alpha, 'OLS')) {
        coefs <- as.matrix(coef(cvfit))
      } else {
        coefs <- coef(cvfit, s = my_lambda)
        # plot(cvfit, main=paste(this_event, cvfit_name, alpha))
      }
      data.table(feature = rownames(coefs), coefs = coefs[, 1])
    }, simplify = FALSE), idcol = 'alpha')
  }, simplify = FALSE), idcol = 'model')
  # as_dt <- melt(dcast(as_dt, model + alpha ~ feature, value.var = 'coefs'), id.vars = c('model', 'alpha'), variable.name = 'feature', value.name = 'coefs')
  # as_dt[, model:=sub('no_enhancer', 'noEnhancer', model, fixed = TRUE)]
  # as_dt[, model:=sub('enhancer_stripped', 'enhancerStripped', model, fixed = TRUE)]
  as_dt[, model:=gsub('nocCRE', 'wo/ cCRE', model, fixed = TRUE)]
  as_dt[, c('PSI Variance', 'dat_expl_resp'):=tstrsplit(model, '.', fixed=TRUE)]
  as_dt[, `PSI Variance`:=tstrsplit(`PSI Variance`, '_', fixed=TRUE, keep = 1)]
  as_dt[, c('dataset', 'explanatory', 'response'):=tstrsplit(dat_expl_resp, '_', fixed=TRUE)]
  # as_dt[, coef_rank:=rank(-coefs, na.last = 'keep'), by=model]
}, simplify = FALSE), idcol = 'Event Type')
split_features(rank_dt, 'feature')
rank_dt[, region:=sub('upstream', 'Upstream', region, fixed = TRUE)]
rank_dt[, region:=sub('downstream', 'Downstream', region, fixed = TRUE)]
rank_dt[, region:=gsub('[_,]', '\n', region)]
rank_dt[, region := factor(
  region,
  levels =
    c(
      unique(region)[startsWith(unique(region), 'Upstream')],
      unique(region)[startsWith(unique(region), 'RI')],
      unique(region)[startsWith(unique(region), 'SE')],
      unique(region)[startsWith(unique(region), 'Downstream')],
      unique(region)[startsWith(unique(region), 'Gene')],
      unique(region)[startsWith(unique(region), 'PLS')],
      unique(region)[startsWith(unique(region), 'DNase-H3K4me3')],
      unique(region)[startsWith(unique(region), 'pELS')],
      unique(region)[startsWith(unique(region), 'dELS')],
      unique(region)[startsWith(unique(region), 'CTCF-only')]
    )
)]
rank_dt[, mark:=sub('gene_expression', 'Gene_Expression', mark)]
rank_dt[, mark:=gsub('[_,]', '\n', mark)]
rank_dt[, mark:=sub('wgbs', 'DNAm', mark)]
rank_dt[, mark:=sub('width', 'Width', mark)]
rank_dt[, mark:=sub('distance', 'Distance', mark)]
rank_dt[, mark := factor(
  mark,
  levels =
    c('(Intercept)', 'ID', 'Gene\nExpression', 'Distance\nTSS', 'Distance\nTES', 'Width', 'DNAm', 'H3K36me3', 'H3K4me1', 'H3K4me3', 'H3K27me3', 'H3K27ac', 'H3K9me3')
)]

rank_dt[, 'Samples, Features':= as.factor(paste(dataset, explanatory, sep = ', '))]
rank_dt[, `PSI Variance`:=factor(`PSI Variance`, levels = c('low', 'all', 'high'))]
rank_dt[, c("random", "alpha"):=tstrsplit(alpha, "::", fixed=TRUE)]
rank_dt[, random := as.logical(random)]
```


```{r, fig.width=16, fig.height=11, eval=FALSE}
# my_colors <- rev(c("#CC79A7", "#009E73", "#999999"))
for (this_event in to_analyze) {
  # pheatmap(as.matrix(dcast(rank_dt[feature != '(Intercept)' & this_event == `Event Type` & alpha == 'OLS' & region != 'Gene-Wide' & mark != 'Width' & grepl('(CTCF)|(DNase)|(ELS)|(PLS)', region)], formula = `Samples, Features` + region ~ `Mark, PSI Variance`, value.var = 'coefs', fill = 0L)[, -c(1, 2)]), cluster_cols = FALSE, cluster_rows = FALSE)
  # print(ggplot(rank_dt[feature != '(Intercept)' & this_event == `Event Type` & alpha == 'OLS' & region != 'Gene-Wide' & mark != 'Width' & grepl('(CTCF)|(DNase)|(ELS)|(PLS)', region)], aes(x = `Samples, Features`, y = coefs, color=`PSI Variance`, shape = `Samples, Features`, size = abs(coefs))) + facet_grid(mark ~ region) + geom_hline(yintercept = 0) + geom_point() + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#element_text(angle = 45, hjust=1))
  #       + labs(y = 'Coefficient', size = 'Absolute Coefficient') + scale_shape_manual(values = LETTERS) + theme(legend.position = 'bottom') + scale_color_manual(values = my_colors))
  # ggsave(filename = file.path(plot_dir, paste0('coef_', this_event, '.pdf')), width = 14, height = 8)
  # print(ggplot(rank_dt[feature != '(Intercept)' & this_event == `Event Type` & alpha == 'OLS' & region != 'Gene-Wide' & mark != 'Width' & !grepl('(CTCF)|(DNase)|(ELS)|(PLS)', region)], aes(x = `Samples, Features`, y = coefs, color=`PSI Variance`, shape = `Samples, Features`, size = abs(coefs))) + facet_grid(mark ~ region) + geom_hline(yintercept = 0) + geom_point() + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#element_text(angle = 45, hjust=1))
  #       + labs(y = 'Coefficient', size = 'Absolute Coefficient') + scale_shape_manual(values = LETTERS) + theme(legend.position = 'bottom') + scale_color_manual(values = my_colors))
  #   ggsave(filename = file.path(plot_dir, paste0('coef_', this_event, '.pdf')), width = 14, height = 8)
  # print(ggplot(rank_dt[feature != '(Intercept)' & this_event == `Event Type` & alpha == 'OLS' & (region == 'Gene-Wide' | mark ==  'Width')], aes(x = `Samples, Features`, y = coefs, color=`PSI Variance`, shape = `Samples, Features`, size = abs(coefs))) + facet_wrap(~ region + mark, nrow = 2) + geom_hline(yintercept = 0) + geom_point() + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#element_text(angle = 45, hjust=1))
  #       + labs(y = 'Coefficient', size = 'Absolute Coefficient') + scale_shape_manual(values = LETTERS) + scale_color_manual(values = my_colors))
  # ggsave(filename = file.path(plot_dir, paste0('coef_non_epigenetic_', this_event, '.pdf')), width = 9, height = 6)
  print(
    ggplot(
      rank_dt[feature != '(Intercept)' &
                this_event == `Event Type` &
                alpha == 'OLS' & !random & (region == 'Gene-Wide' | mark ==  'Width')], 
    aes(
        y = `Samples, Features`,
        x = `PSI Variance`,
        fill = coefs
    )
    ) + facet_wrap( ~ region + mark, nrow = 2) + geom_tile() + theme_bw() + scale_fill_gradient2() + scale_y_discrete(limits=rev) + theme(legend.position = 'bottom')
    #just_text(angle = 45, hjust=1))+labs(y = 'Coefficient', size = 'Absolute Coefficient') + scale_shape_manual(values = LETTERS) + scale_color_manual(values = my_colors)
  )
  ggsave(filename = file.path(plot_dir, paste0('coef_non_epigenetic_', this_event, '.pdf')), width = 5, height = 6)
  print(
    ggplot(
    rank_dt[feature != '(Intercept)' &
                this_event == `Event Type` &
                alpha == 'OLS' &
                region != 'Gene-Wide' &
                mark != 'Width' &
                grepl('(CTCF)|(DNase)|(ELS)|(PLS)', region)],
    aes(
        y = `Samples, Features`,
        x = `PSI Variance`,
        fill = coefs
    )
    ) + facet_grid(region ~ mark, scales = 'free_y') + geom_tile() + theme_bw() + scale_fill_gradient2() + scale_y_discrete(limits=rev)
  #just_text(angle = 45, hjust=1))+labs(y = 'Coefficient', size = 'Absolute Coefficient') + scale_shape_manual(values = LETTERS) + scale_color_manual(values = my_colors)
  )
  ggsave(filename = file.path(plot_dir, paste0('coef_cCRE_', this_event, '.pdf')), width =9, height = 10.5)
  print(
    ggplot(
    rank_dt[feature != '(Intercept)' &
                this_event == `Event Type` &
                alpha == 'OLS' &
                region != 'Gene-Wide' &
                mark != 'Width' &
                !grepl('(CTCF)|(DNase)|(ELS)|(PLS)', region)],
    aes(
        y = `Samples, Features`,
        x = `PSI Variance`,
        fill = coefs
    )
    ) + facet_grid(region ~ mark) + geom_tile() + theme_bw() + scale_fill_gradient2() + scale_y_discrete(limits=rev)
    #just_text(angle = 45, hjust=1))+labs(y = 'Coefficient', size = 'Absolute Coefficient') + scale_shape_manual(values = LETTERS) + scale_color_manual(values = my_colors)
  )
  ggsave(filename = file.path(plot_dir, paste0('coef_exon_intron', this_event, '.pdf')), width =9, height = 7)
  
}
```

# make roc_curves for total models:
```{r, fig.height=6, fig.width=12, eval=FALSE}
# library(ggplot2)
# library(glmnet)
# library(ranger)
library(PRROC)
library(patchwork)
all_plots <- list()
all_aucs <- list()
cm_list <- sapply(to_analyze, function(this_event){
    feature_vector <- c('noEpigenetic', 'all', 'onlyEpigenetic')
    feature_names <- c('Intrinsic', 'Epigenetic and Intrinsic', 'Epigenetic')
    names(feature_names) <- feature_vector
    vars <- c('low_var', 'high_var', 'all_var')
    var_names <- c('Low Variability', 'High Variability', 'All Variability')
    names(var_names) <- vars
    combs <- expand.grid(feature_vector=feature_vector, vars=vars)
    as_list <- pbmcmapply(features=combs$feature_vector, var=combs$vars, FUN = function(features, var) {
        models <- readRDS(paste(this_event, var, features, 'binomial.rds', sep = '_'))
        print(paste(this_event, var, features))
        cvfits <- models$cvfit
        explanatory <- models$explanatory
        # real_test_ids <-
        #   Reduce(intersect, lapply(names(models$test_by_group), function(grouping_col)
        #     test_data[get(grouping_col) %in% models$test_by_group[[grouping_col]], which = TRUE]))
        only_first <- models$test_data[get(names(models$test_by_group)[1]) %in% models$test_by_group[[1]] &
                           !(get(names(models$test_by_group)[2]) %in% models$test_by_group[[2]])]
        only_second <- models$test_data[get(names(models$test_by_group)[2]) %in% models$test_by_group[[2]] &
                           !(get(names(models$test_by_group)[1]) %in% models$test_by_group[[1]])]
        strict_test_data <- models$test_data[do.call(CJ, models$test_by_group), on=names(models$test_by_group), nomatch=NULL]
        test_data_list <- list(only_first, only_second, strict_test_data)
        names(test_data_list) <- c(names(models$test_by_group), paste(names(models$test_by_group), collapse = "_AND_"))
        test_res <- sapply(test_data_list, 
               function(test_data){
          sapply(names(cvfits), function(alpha){
            cvfit <- cvfits[[alpha]]
            
            if (endsWith(alpha, 'OLS')) {
              predicted <-
                predict(cvfit, newdata = test_data[, ..explanatory], type = 'response')
              predicted_label <- ifelse(predicted < .5, control_class, case_class)
              varImpDT <- as.data.table(summary(cvfit)[["coefficients"]], keep.rownames = 'feature')
            } else {
              predicted <-
                predict(cvfit, newdata = test_data[, ..explanatory], type = 'prob')[, case_class]
              predicted_label <- predict(cvfit, newdata = test_data[, ..explanatory], type = 'raw')
              if (endsWith(alpha, 'RF')) {
                varImpDT <- as.data.table(varImp(cvfit)$importance, keep.rownames = 'feature')
                setnames(varImpDT, 'Overall', 'Estimate')
              } else { # glmnet 
                varImpDT <- as.data.table(as.matrix(coef(cvfit$finalModel, s = cvfit$bestTune$lambda)), keep.rownames = 'feature')
                setnames(varImpDT, 's1', 'Estimate')
              }
            }
          predicted_label <- ordered(predicted_label, levels = class_levels)
          roc <- PRROC::roc.curve(scores.class0 = predicted, weights.class0 = test_data[, ifelse(binary == case_class, 1, 0)], curve = TRUE)
          if (length(unique(predicted)) > 1)
            prc <- PRROC::pr.curve(scores.class0 = predicted, weights.class0 = test_data[, ifelse(binary == case_class, 1, 0)], curve = TRUE)
          else {
            prc <- PRROC::pr.curve(scores.class0 = predicted, weights.class0 = test_data[, ifelse(binary == case_class, 1, 0)], curve = FALSE)
            prc$curve <- data.table(c(0, 1), prc$auc.integral, unique(predicted))
          }
            browser()
          list(AUROC=roc$auc,
               AUPRC=prc$auc.integral,
               dt=setNames(rbind(data.table(roc$curve)[, type:=roc$type], data.table(prc$curve)[, type:=prc$type]), c('Recall', 'Precision', 'Threshold', 'Type')),
               cm=caret::confusionMatrix(predicted_label, test_data[, binary], positive=case_class, mode = "everything"),
               MCC=ModelMetrics::mcc(test_data[, ifelse(binary == case_class, 1, 0)], predicted, 0.5),
               size_dt=data.table(nsamples=models$nsamples, npos_train=test_data[, sum(binary == case_class)], ntest=test_data[, .N], npos_test=test_data[, sum(binary == case_class)]),
               abline_dt=data.table(intercept=c(0, test_data[, sum(binary == case_class)/.N]), slope=c(1, 0), Type=c('ROC', 'PR')),
               varImpDT=varImpDT
               )
          }, simplify = FALSE)
        }, simplify = FALSE)
        names(test_res)[names(test_res) == 'seqnames'] <- 'Unknown Chromosome'
        names(test_res)[names(test_res) == 'harmonized_sample_ontology_term_high_order_fig1'] <- 'Unknown Tissue/Cell'
        names(test_res)[names(test_res) == 'seqnames_AND_harmonized_sample_ontology_term_high_order_fig1'] <- 'Both Unknown'
        
        test_res
    }, SIMPLIFY = FALSE)
    names(as_list) <- paste(feature_names[combs$feature_vector], var_names[combs$vars], sep = "::")
    varImp_dt <- rbindlist(lapply(as_list, function(test_list) rbindlist(lapply(test_list, function(roc_list) rbindlist(lapply(roc_list, function(element) element$varImpDT), idcol = 'Model', fill = TRUE)), idcol = 'Test on', fill = TRUE)), idcol = 'Features', fill = TRUE)
    curve_dt <- rbindlist(lapply(as_list, function(test_list) rbindlist(lapply(test_list, function(roc_list) rbindlist(lapply(roc_list, function(element) element$dt), idcol = 'Model')), idcol = 'Test on')), idcol = 'Features')
    abline_dt <- rbindlist(lapply(as_list, function(test_list) rbindlist(lapply(test_list, function(roc_list) rbindlist(lapply(roc_list, function(element) element$abline_dt), idcol = 'Model')), idcol = 'Test on')), idcol = 'Features')
    varImp_dt[, c('Features', 'Variability') := tstrsplit(Features, '::')]
    fwrite(varImp_dt, paste0(this_event, '_varImp.csv'))
    abline_dt[, c('Features', 'Variability') := tstrsplit(Features, '::')]
    curve_dt[, c('Features', 'Variability') := tstrsplit(Features, '::')]
    curve_dt[, (c('Random', 'Model')):=tstrsplit(Model, '::', fixed = TRUE)]
    curve_dt[Random == 'FALSE', Random:="Real"]
    curve_dt[Random == 'TRUE', Random:="Random"]
    print(ggplot(curve_dt, aes(x=Recall, y=Precision, color=Model, linetype=Features)) + geom_abline(data=abline_dt, aes(slope=slope, intercept=intercept), color='grey') + geom_line() + facet_grid(Type + Random ~ `Test on` + Variability) + theme_bw() + labs(title = this_event))
    measure_dt <- rbindlist(lapply(as_list, function(test_list) 
      rbindlist(lapply(test_list, function(roc_list)
      rbindlist(lapply(roc_list, function(element)
        c(
          element[c('AUROC', 'AUPRC', 'MCC')],
          list(
            Kappa = element$cm$overall['Kappa'],
            `Balanced\nAccuracy` = element$cm$byClass['Balanced Accuracy']
          )
        )), idcol = 'Model')), idcol = 'Test on')), idcol = 'Features')
    measure_dt <- melt(measure_dt, id.vars = c('Features', 'Model', 'Test on'), variable.name = 'Measure', value.name = 'Value')
    measure_dt[, (c('Labels', 'Model')):=tstrsplit(Model, '::', fixed = TRUE)]
    measure_dt[Labels == 'FALSE', Labels:="Real"]
    measure_dt[Labels == 'TRUE', Labels:="Random"]
    measure_dt[, c('Features', 'Variability') := tstrsplit(Features, '::')]
  print(ggplot(measure_dt, aes(y = Value, x = Model, fill = Features, color = Labels, alpha = Labels)) + geom_col(position = 'dodge') + facet_grid(Measure ~ `Test on` + Variability) + scale_color_manual(values=c('Real'='black', 'Random'='NA')) + scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) + theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust=1)) + labs(title = this_event))
    list(measure_dt=measure_dt, curve_dt=curve_dt, abline_dt=abline_dt, size_table=rbindlist(lapply(as_list, function(test_list) rbindlist(lapply(test_list, function(roc_list) rbindlist(lapply(roc_list, function(element) element$size_dt), idcol = 'Model')), idcol = 'Test on')), idcol = 'Features'))
  # flat_as_list <- unlist(unlist(as_list, recursive = FALSE), recursive = FALSE)
  # names(flat_as_list) <- gsub('nocCRE', 'wo/ cCRE', names(flat_as_list), fixed = TRUE)
  # for (alpha in names(models[[1]]$cvfit)) {
  #   roc_list <- flat_as_list[endsWith(names(flat_as_list), paste0(alpha, '.roc'))]
  #   roc_list <- roc_list[order(names(roc_list))]
  #   auc_list <- sapply(names(roc_list), function(x){
  #     var <- tstrsplit(x, '_', fixed = TRUE, keep = 1)[[1]]
  #     paste0(var, ': ', round(roc_list[[x]][["auc"]], 2))
  #   })
  # plot_list <- lapply(unique(tstrsplit(names(roc_list), '.', fixed=TRUE, keep=1)[[1]]), function(var) {
  #     tmp_roc_list <- roc_list[startsWith(names(roc_list), var)]
  #     if (var == 'all_var' && all(grepl('FALSE::RF', names(roc_list), fixed=TRUE))) {
  #       all_aucs[[this_event]] <<- sub('all', this_event, auc_list[names(tmp_roc_list)])
  #       all_aucs[[paste(this_event, "simple", sep = "_")]] <<- sub('all', this_event, auc_list[paste0(var,'.', c("all_all", "wo/ cCRE_all"), '_binomial.', alpha, '.roc')])
  #     }
  #     names(tmp_roc_list) <- gsub(paste0('_binomial\\.', alpha, '\\.roc$'), '', gsub(paste0('^',var,'\\.'), '', names(tmp_roc_list)))
  #     
  #     if (var == 'all_var' && all(grepl('FALSE::RF', names(roc_list), fixed=TRUE))) {
  #     all_plots[[this_event]] <<- ggroc(tmp_roc_list) + geom_abline(intercept = 1, slope = 1,
  #               color = "darkgrey", linetype = "dashed") +
  #           labs(title=this_event, color='Samples, Features (Event Type: AUC)', x = 'Specificity', y = 'Sensitivity') +
  #           theme_bw() + theme(legend.position = 'bottom') + guides(color=guide_legend(nrow=4))
  #     all_plots[[paste(this_event, "simple", sep = "_")]] <<- ggroc(tmp_roc_list[c("all_all", "wo/ cCRE_all")]) + geom_abline(intercept = 1, slope = 1,
  #               color = "darkgrey", linetype = "dashed") +
  #           labs(title=this_event, color='Samples, Features (Event Type: AUC)', x = 'Specificity', y = 'Sensitivity') +
  #           theme_bw() + theme(legend.position = 'bottom')
  #     }
  #     
  #     p <- ggroc(tmp_roc_list) + geom_abline(intercept = 1, slope = 1,
  #               color = "darkgrey", linetype = "dashed") +
  #           labs(title=tstrsplit(var, '_', fixed = TRUE, keep = 1)[[1]], color='Samples, Features (variance: AUC)', x = 'Specificity', y = 'Sensitivity') + 
  #           scale_color_discrete(labels=sapply(names(tmp_roc_list), function(x) paste0(sub('_', ', ', x, fixed = TRUE), ' (', paste(auc_list[grep(x, names(auc_list), fixed = TRUE)], collapse = ', '), ')'))) + 
  #           theme_bw() + theme(legend.position = 'bottom') + guides(color=guide_legend(nrow=4))
  #     p
  #   })
  # print(Reduce(`+`, plot_list) + plot_layout(guides = "collect") + plot_annotation(paste(this_event, alpha)) & theme(legend.position = 'bottom'))# + plot_layout(guides = "collect") & theme(legend.position = 'bottom'))
  # # ggsave(filename = file.path(plot_dir, paste0('roc_', this_event, '_', alpha, '.pdf')), width = 11.7, height = 5)
  #   # print(ggroc(roc_list) + 
  #         # labs(title=paste(this_event, alpha), color='data_explanatory_response_alpha') + scale_color_discrete(labels=paste(names(roc_list), sapply(roc_list, function(x) paste0('(', round(x[["auc"]], 2),')')))) + theme_bw())
  # }
  # 
  # flat_as_list[endsWith(names(flat_as_list), '.cm')]
}, simplify = FALSE)
gc()
saveRDS(cm_list, 'cm_list.rds')
```
```{r}
cm_list <- readRDS('cm_list.rds')
```

```{r}
for (this_event in to_analyze) {
  print(cm_list[[this_event]]$size_table[, .(Event=this_event, Models=paste(unique(paste(tstrsplit(x = Features, split='::', fixed=TRUE)[[2]], `Test on`, sep = '::')), collapse = ', ')), by=.(nsamples, ntest, npos_test/ntest)])  
  varImp_dt <- fread(paste0(this_event, '_varImp.csv'), stringsAsFactors = TRUE)
  varImp_dt[, feature := gsub(pattern = '`', replacement = '', x = feature, fixed = TRUE)]
}
```


```{r}
curve_dt <- rbindlist(sapply(cm_list, function(x) x$curve_dt, simplify=FALSE), idcol = 'Event Type')
curve_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
curve_dt[Features == 'Intrinsic', Features := 'Non Epigenetic']
curve_dt[Features == 'Epigenetic and Intrinsic', Features := 'All']
curve_dt[, Features:= factor(Features, levels = c('All', 'Non Epigenetic', 'Epigenetic'))]
curve_dt[, Response:=factor(Random, levels=c('Real', 'Random'))]
abline_dt <- rbindlist(sapply(cm_list, function(x) x$abline_dt, simplify=FALSE), idcol = 'Event Type')
abline_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
abline_dt[, Features := factor(Features, levels = c('Epigenetic and Intrinsic', 'Intrinsic', 'Epigenetic'))]
abline_dt[Features == 'Intrinsic', Features := 'Non Epigenetic']
abline_dt[Features == 'Epigenetic and Intrinsic', Features := 'All']
abline_dt[, Features:= factor(Features, levels = c('All', 'Non Epigenetic', 'Epigenetic'))]

ggplot(curve_dt[Model == 'RF' & `Test on` == 'Both Unknown'], 
             aes(x=Recall, y=Precision, color=Features, linetype=Response)) + 
        geom_abline(data=unique(abline_dt[`Test on` == 'Both Unknown', .(Variability, `Event Type`, intercept, slope, Type)]), aes(slope=slope, intercept=intercept), color='black') + 
        geom_line() + 
        ggh4x::facet_grid2(Type ~ `Event Type` + Variability, strip = ggh4x::strip_nested()) + 
        scale_y_continuous(breaks = c(0, 1)) +
        scale_x_continuous(breaks = c(0, 1)) +
        theme_bw() + theme(strip.background = element_rect(fill = 'white'), legend.position = 'bottom')
ggsave(filename = file.path(plot_dir, 'MLcurves.pdf'), width = 9, height = 4)
ggsave(filename = file.path(publication_plots_dir, 'MLcurves.pdf'), width = 9, height = 4)
```


```{r}
mixed_metrics <- rbindlist(sapply(cm_list, function(x) x$measure_dt, simplify=FALSE), idcol = 'Event Type')

mixed_metrics[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
mixed_metrics[Features == 'Intrinsic', Features := 'Non Epigenetic']
mixed_metrics[Features == 'Epigenetic and Intrinsic', Features := 'All']
mixed_metrics[, Features:= factor(Features, levels = c('All', 'Non Epigenetic', 'Epigenetic'))]
mixed_metrics[, Measure := factor(Measure, levels = c('Balanced\nAccuracy', 'AUROC', 'AUPRC', 'MCC', 'Kappa'))]
mixed_metrics[, Variability := factor(Variability, levels = c('High Variability', 'All Variability', 'Low Variability'))]
mixed_metrics[, Response:=factor(Labels, levels=c('Random','Real'))]

mixed_metrics[`Test on` == 'Both Unknown', `Test on` := 'Chr &\nCell']
mixed_metrics[`Test on` == 'Unknown Chromosome', `Test on` := 'Chr']
mixed_metrics[`Test on` == 'Unknown Tissue/Cell', `Test on` := 'Cell']
# mixed_metrics[, `Test on`:=sub(x = `Test on`, pattern = ' ', replacement = '\n', fixed = TRUE)]
mixed_metrics[, `Test on`:=factor(`Test on`, levels=c('Chr &\nCell', 'Chr', 'Cell'))]

hline_dt <- setDT(mixed_metrics[, expand.grid(`Event Type`=levels(`Event Type`), 'Variability'=levels(Variability), Measure=levels(Measure))])
hline_dt[copy(unique(abline_dt[`Test on` == 'Both Unknown' & Type == 'PR', .(Variability, `Event Type`, intercept, slope, Type)]))[, Measure := 'AUPRC'], on=.NATURAL, intercept:=intercept]
hline_dt[Measure == 'AUROC', intercept := 0.5]
hline_dt[Measure == 'Balanced\nAccuracy', intercept := 0.5]
hline_dt[Measure == 'MCC', intercept := 0]
hline_dt <- hline_dt[Measure != 'Kappa']

ggplot(mixed_metrics[Model == 'RF' & `Test on` == 'Chr &\nCell' & Measure != 'Kappa'],
       aes(y = Value, x = Variability, fill = Features, color = Response, alpha = Response)) +
  geom_col(position = 'dodge') +
  facet_grid(Measure ~ `Event Type`) +
  scale_color_manual(values=c('Real'='black', 'Random'='NA')) +
  scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) +
  theme_bw() + theme(strip.background = element_rect(fill = 'white')) +
  geom_text(aes(label = paste0('.', 100*round(Value, 2)), angle=90, hjust=-0.1, vjust=0.5, color='black'), show.legend = FALSE, position=position_dodge(width=0.9)) +
  scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = .25)) +
  theme(legend.position = 'bottom')
ggsave(filename = file.path(plot_dir, 'mixed_metrics.pdf'), width = 10, height = 5)
ggplot(mixed_metrics[Model == 'RF' & `Test on` == 'Chr &\nCell' & Measure != 'Kappa'], 
       aes(y = Value, x = Labels, color = Features)) + 
    geom_hline(data=hline_dt, aes(yintercept=intercept), linetype='dashed', color = 'grey') +
    geom_point() +
    geom_line(aes(group = Features)) +
    ggh4x::facet_grid2(Measure ~ `Event Type` + Variability, strip = ggh4x::strip_nested(), scales='free') + 
    scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) + 
    theme_bw() + theme(strip.background = element_rect(fill = 'white')) + 
    # expand_limits(y = 1) +
    # scale_y_continuous(limits = c(mixed_metrics[Model == 'RF' & `Test on` == 'Both Unknown' & Measure != 'Kappa', min(Value)], 1)
    scale_y_continuous(breaks = function(x) seq(ceiling(min(x) * 10)/10, floor(max(x) * 10)/10, by = .2)) +
    theme(legend.position = 'bottom')
ggsave(filename = file.path(plot_dir, 'mixed_metrics_dumbbell.pdf'), width = 6.5, height = 4)
ggsave(filename = file.path(publication_plots_dir, 'mixed_metrics_dumbbell.pdf'), width = 6.5, height = 4)
```
```{r}
mixed_metrics[Model == "1", Model:='LASSO']
mixed_metrics[Model == "OLS", Model:='LogReg']
mixed_metrics[, Model:=factor(Model, levels=c('LASSO', 'LogReg', 'RF'))]
ggplot(mixed_metrics[`Test on` == 'Chr &\nCell' & Measure != 'Kappa' & Labels == 'Real'],
       aes(y = Value, x = Model, color = Features)) + 
    geom_hline(data=hline_dt, aes(yintercept=intercept), linetype='dashed', color = 'grey') +
    geom_point() +
    geom_line(aes(group = Features)) +
    ggh4x::facet_grid2(Measure ~ `Event Type` + Variability, strip = ggh4x::strip_nested(), scales='free') + 
    scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) + 
    theme_bw() + theme(strip.background = element_rect(fill = 'white'), axis.text.x = element_text(angle=45, hjust=1)) + 
    # expand_limits(y = 1) +
    # scale_y_continuous(limits = c(mixed_metrics[Model == 'RF' & `Test on` == 'Both Unknown' & Measure != 'Kappa', min(Value)], 1)
    scale_y_continuous(breaks = function(x) seq(ceiling(min(x) * 10)/10, floor(max(x) * 10)/10, by = .2)) +
    theme(legend.position = 'none')
ggsave(filename = file.path(plot_dir, 'mixed_metrics_model_dumbbell.pdf'), width = 6.5, height = 4)
ggsave(filename = file.path(publication_plots_dir, 'mixed_metrics_model_dumbbell.pdf'), width = 6.5, height = 4)
```

```{r}
# TODO double check hline_dt. AUPRC is there twice for each combination
hline_dt <- setDT(mixed_metrics[, expand.grid(`Event Type`=levels(`Event Type`), Variability=levels(Variability), Measure=levels(Measure), `Test on`=levels(`Test on`))])
hline_dt[copy(unique(abline_dt[Type == 'PR', .(Variability, `Event Type`, intercept, slope, Type)]))[, Measure := 'AUPRC'], on=.NATURAL, intercept:=intercept]
hline_dt[Measure == 'AUROC', intercept := 0.5]
hline_dt[Measure == 'Balanced\nAccuracy', intercept := 0.5]
hline_dt[Measure == 'MCC', intercept := 0]
hline_dt <- hline_dt[Measure != 'Kappa']

ggplot(mixed_metrics[Model == 'RF' & Labels == 'Real' & Measure != 'Kappa'], 
       aes(y = Value, x = `Test on`, color = Features)) + 
    geom_hline(data=hline_dt, aes(yintercept=intercept), linetype='dashed', color = 'grey') +
    geom_point() +
    geom_line(aes(group = Features)) +
    ggh4x::facet_grid2(Measure ~ `Event Type` + Variability, strip = ggh4x::strip_nested(), scales='free') + 
    scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) + 
    theme_bw() + theme(strip.background = element_rect(fill = 'white'), axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) + 
    # expand_limits(y = 1) +
    # scale_y_continuous(limits = c(mixed_metrics[Model == 'RF' & `Test on` == 'Both Unknown' & Measure != 'Kappa', min(Value)], 1), breaks = seq(0, 1, by = .5)) +
    scale_y_continuous(breaks = function(x) seq(ceiling(min(x) * 10)/10, floor(max(x) * 10)/10, by = .2)) +
    labs(x = "Test on Unseen Chromosome (Chr) or Cell Type (Cell) or Both (Chr & Cell)") +
    theme(legend.position = 'bottom')
ggsave(filename = file.path(plot_dir, 'mixed_metrics_test_dumbbell.pdf'), width = 6.5, height = 5)
ggsave(filename = file.path(publication_plots_dir, 'mixed_metrics_test_dumbbell.pdf'), width = 6.5, height = 5)
```

```{r}
library(patchwork)
library(ggh4x)
varImp_dt <- rbindlist(sapply(to_analyze, function(this_event) fread(paste0(this_event, '_varImp.csv'), stringsAsFactors = TRUE), simplify = FALSE), idcol = 'Event Type')

varImp_dt[, `Event Type`:=factor(`Event Type`, levels=c('SE', 'RI'))]
varImp_dt[`Event Type` == 'SE', feature := gsub('other_region', 'Intron', feature, fixed=TRUE)]
varImp_dt[`Event Type` == 'RI', feature := gsub('other_region', 'Exon', feature, fixed=TRUE)]
varImp_dt[`Event Type` == 'SE', feature := gsub('event_name', 'SE', feature, fixed=TRUE)]
varImp_dt[`Event Type` == 'RI', feature := gsub('event_name', 'RI', feature, fixed=TRUE)]

varImp_dt[, let(c('Random', 'Architecture'), tstrsplit(Model, '::', fixed=TRUE))]
# varImp_dt[, feature := gsub(pattern = '`', replacement = '', x = feature, fixed = TRUE)]
varImp_dt[, feature:=gsub('_', ' ', feature, fixed = TRUE)]
varImp_dt[, let(c('feature', 'region'), tstrsplit(feature, ';', fixed=TRUE))]
varImp_dt[, region:=tools::toTitleCase(region)]
varImp_dt[, feature:=tools::toTitleCase(feature)]
varImp_dt[, region:=sub(pattern = ' ', replacement = '\n', x = region, fixed = TRUE)]
# varImp_dt[, feature:=sub(pattern = ' ', replacement = '\n', x = feature, fixed = TRUE)]
varImp_dt[Features == 'Intrinsic', Features := 'Non Epigenetic']
varImp_dt[Features == 'Epigenetic and Intrinsic', Features := 'All']
varImp_dt[, Features:= factor(Features, levels = c('All', 'Non Epigenetic', 'Epigenetic'))]
# varImp_dt[, Features := factor(Features, levels = c('Epigenetic and Intrinsic', 'Intrinsic', 'Epigenetic'))]

region_plot <- ggplot(varImp_dt[Architecture == 'RF' & `Test on` == 'Both Unknown' & Random == 'FALSE' & !is.na(region)], aes(x = factor(region, levels = c("Upstream\nIntron", "SE", "Downstream\nIntron", "Upstream\nExon", "RI", "Downstream\nExon")), 
                                                                                                                              y = reorder(feature, Estimate), 
                                                                                                                              fill = Estimate)) + 
    geom_tile() + 
    ggh4x::facet_grid2(Features ~ `Event Type` + Variability, scales = 'free', strip = ggh4x::strip_nested()) +
    theme_bw() + 
    theme(strip.background = element_rect(fill = 'white')) + 
    scale_fill_continuous(type = 'viridis') + 
    labs(y = 'Region-dependent Features', x = 'Region', fill = 'Variable\nImportance')

no_region_plot <- ggplot(varImp_dt[Architecture == 'RF' & `Test on` == 'Both Unknown' & Random == 'FALSE' & is.na(region)], aes(x = region, y = reorder(feature, Estimate), fill = Estimate)) + 
    geom_tile() + 
    ggh4x::facet_grid2(Features ~ `Event Type` + Variability, scales = 'free', strip = ggh4x::strip_nested()) +
    theme_bw() + 
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), strip.background = element_rect(fill = 'white')) + 
    scale_fill_continuous(type = 'viridis') + 
    labs(y = 'Region-independent Features', x = NULL, fill = 'Variable\nImportance')

full_plot <- (no_region_plot / region_plot + plot_layout(guides = "collect", axis_titles = "collect"))
print(full_plot)
ggsave(full_plot, filename = file.path(plot_dir, 'varImp.pdf'), width = 12, height = 8)
ggsave(full_plot, filename = file.path(publication_plots_dir, 'varImp.pdf'), width = 12, height = 8)
```


```{r, eval=FALSE}
poster_dt <- mixed_metrics[Model == 'RF' & `Test on` == 'Both Unknown' & Measure %in% c('Balanced\nAccuracy', 'AUPRC', 'MCC') & Variability == 'All Variability']
poster_dt[Features == 'Intrinsic', Features := 'Non Epigenetic']
poster_dt[Features == 'Epigenetic and Intrinsic', Features := 'All']
poster_dt[, Features:= factor(Features, levels = c('All', 'Non Epigenetic', 'Epigenetic'))]
ggplot(poster_dt, aes(y = Value, x = Variability, fill = Features, color = Labels, alpha = Labels)) + geom_col(position = 'dodge') + facet_grid(Measure ~ `Event Type`) + scale_color_manual(values=c('Real'='black', 'Random'='NA')) + scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) + theme_bw() + theme(strip.background = element_rect(fill = 'white')) + geom_text(aes(label = paste0('.', 100*round(Value, 2)), angle=90, hjust=-0.1, vjust=0.5, color='black'), show.legend = FALSE, position=position_dodge(width=0.9)) + scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = .25)) + theme(legend.position = 'bottom', axis.title.x=element_blank(), axis.text.x=element_blank(), axis.title.y=element_blank(), axis.ticks.x=element_blank())
ggsave(filename = file.path(plot_dir, 'poster_metrics.svg'), width = 7, height = 3.5)
```


```{r, eval=FALSE}
ggplot(mixed_metrics[Model == 'RF' & `Test on` == 'Both Unknown' & Measure %in% c('Balanced\nAccuracy', 'AUPRC', 'MCC') & Variability == 'All Variability' & `Event Type` == 'SE'], aes(y = Value, x = Features, fill = Features, color = Labels, alpha = Labels)) + geom_col(position = 'dodge') + facet_grid(Measure ~ `Event Type`) + scale_color_manual(values=c('Real'='black', 'Random'='NA')) + scale_alpha_manual(values=c('Real'=1, 'Random'=0.5)) + theme_bw() + theme(strip.background = element_rect(fill = 'white')) + geom_text(aes(label = paste0('.', 100*round(Value, 2)), angle=90, hjust=-0.1, vjust=0.5, color='black'), show.legend = FALSE, position=position_dodge(width=0.9)) + scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = .25)) + theme(legend.position = 'bottom', axis.title.y=element_blank()) + guides(fill="none")
ggsave(filename = file.path(plot_dir, 'poster_metrics_SE.pdf'), width = 4, height = 4)
```


```{r, eval=FALSE}
((all_plots$SE + scale_color_discrete(labels=paste0(sub('_', ', ', gsub('_binomial\\.FALSE::RF\\.roc$', '', gsub('^all_var\\.', '', names(all_aucs$SE)))), ' (', all_aucs$SE, ' , ', all_aucs$RI , ')'))) + (all_plots$RI + scale_color_discrete(labels=paste0(sub('_', ', ', gsub('_binomial\\.FALSE::RF\\.roc$', '', gsub('^all_var\\.', '', names(all_aucs$SE)))), ' (', all_aucs$SE, ' , ', all_aucs$RI , ')')))) + plot_layout(guides = "collect") & theme(legend.position = 'bottom')
ggsave(filename = file.path(plot_dir, 'roc_both.pdf'), width = 9, height = 5.5)
print(cm_list$SE$`all_var.all_all_binomial.FALSE::RF.cm`)
print(cm_list$RI$`all_var.all_all_binomial.FALSE::RF.cm`)
```
```{r figure6, eval=FALSE}
((all_plots$SE_simple + scale_color_discrete(labels=paste0(sub('_', ', ', gsub('_binomial\\.FALSE::RF\\.roc$', '', gsub('^all_var\\.', '', names(all_aucs$SE_simple)))), ' (', all_aucs$SE_simple, ' , ', all_aucs$RI_simple , ')'))) + (all_plots$RI_simple + scale_color_discrete(labels=paste0(sub('_', ', ', gsub('_binomial\\.FALSE::RF\\.roc$', '', gsub('^all_var\\.', '', names(all_aucs$SE_simple)))), ' (', all_aucs$SE_simple, ' , ', all_aucs$RI_simple, ')')))) + plot_layout(guides = "collect") & theme(legend.position = 'bottom')
ggsave(filename = file.path(plot_dir, 'roc_both_simple.pdf'), width = 9, height = 5.5)
```
 
```{r, eval=FALSE}
size_table <- rbindlist(sapply(to_analyze, function(this_event) rbindlist(lapply(all_models[, this_event], function(model){data.table('Train Size' = model$nsamples, 'Test Size' = nrow(model$test_data))}), idcol='model'), simplify = FALSE), idcol = 'event_name')
size_table[, var := tstrsplit(model, '_', fixed=TRUE, keep=1)]
size_table[, data := tstrsplit(model, '.', fixed=TRUE, keep=2)]
size_table[, data := tstrsplit(data, '_', fixed=TRUE, keep=1)]
size_table <- unique(size_table, by=c('event_name', 'Train Size', 'Test Size', 'var', 'data'))
size_table_cast <- dcast(size_table, var + data ~ event_name, value.var = c('Train Size', 'Test Size'))
fwrite(size_table_cast[, c(1,2,4,6,3,5)], 'ml_data_sizes.txt')
size_table_cast
```

